<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="SmCCNet">
<title>Reconstructing Phenotype-Specific Multi-Omics Networks with SmCCNet • SmCCNet</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Reconstructing Phenotype-Specific Multi-Omics Networks with SmCCNet">
<meta property="og:description" content="SmCCNet">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">SmCCNet</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/SmCCNet_Vignette_AutoSmCCNet.html">Automated SmCCNet</a>
    <a class="dropdown-item" href="../articles/SmCCNet_Vignette_MultiOmics.html">Reconstructing Phenotype-Specific Multi-Omics Networks with SmCCNet</a>
    <a class="dropdown-item" href="../articles/SmCCNet_Vignette_SingleOmics.html">Reconstructing Phenotype-Specific Single-Omics Networks with SmCCNet</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/KechrisLab/SmCCNet/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Reconstructing Phenotype-Specific Multi-Omics Networks with SmCCNet</h1>
                        <h4 data-toc-skip class="author">Weixuan
Liu</h4>
                        <h4 data-toc-skip class="author">Yonghua
Zhuang</h4>
                        <h4 data-toc-skip class="author">W. Jenny
Shi</h4>
                        <h4 data-toc-skip class="author">Katerina
Kechris</h4>
            
            <h4 data-toc-skip class="date">2024-01-14</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/KechrisLab/SmCCNet/blob/HEAD/vignettes/SmCCNet_Vignette_MultiOmics.Rmd" class="external-link"><code>vignettes/SmCCNet_Vignette_MultiOmics.Rmd</code></a></small>
      <div class="d-none name"><code>SmCCNet_Vignette_MultiOmics.Rmd</code></div>
    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>Sparse multiple canonical correlation network analysis
      (SmCCNet) is a machine learning technique for integrating omics
      data along with a variable of interest (e.g., phenotype of complex
      disease), and reconstructing multiomics networks that are specific
      to this variable. We present the second-generation SmCCNet
      (SmCCNet 2.0) that adeptly integrates single or multiple omics
      data types along with a quantitative or binary phenotype of
      interest. In addition, this new package offers a streamlined setup
      process that can be configured manually or automatically, ensuring
      a flexible and user-friendly experience.</p>
    </div>
    
<div class="section level2">
<h2 id="smccnet-overview">SmCCNet overview<a class="anchor" aria-label="anchor" href="#smccnet-overview"></a>
</h2>
<p><strong>Note:</strong> if you use SmCCNet in published research,
please cite:</p>
<blockquote>
<p>Liu, W., Vu, T., Konigsberg, I. R., Pratte, K. A., Zhuang, Y., &amp;
Kechris, K. J. (2023). SmCCNet 2.0: an Upgraded R package for
Multi-omics Network Inference. bioRxiv, 2023-11.</p>
</blockquote>
<blockquote>
<p>Shi, W. J., Zhuang, Y., Russell, P. H., Hobbs, B. D., Parker, M. M.,
Castaldi, P. J., … &amp; Kechris, K. (2019). Unsupervised discovery of
phenotype-specific multi-omics networks. Bioinformatics, 35(21),
4336-4343.</p>
</blockquote>
<div class="section level3">
<h3 id="workflow">Workflow<a class="anchor" aria-label="anchor" href="#workflow"></a>
</h3>
<p>SmCCNet is a canonical correlation based integration method that
reconstructs phenotype-specific multi-omics networks (Shi et al., 2019).
The algorithm is based on sparse multiple canonical analysis (SmCCA) for
<span class="math inline">\(T\)</span> omics data <span class="math inline">\(X_1, X_2,...X_T\)</span> and a quantitative
phenotype <span class="math inline">\(Y\)</span> measured on the same
subjects. SmCCA finds the canonical weights <span class="math inline">\(w_1, w_2,...,w_T\)</span> that maximize the
(weighted or unweighted) sum of pairwise canonical correlations between
<span class="math inline">\(X_1, X_2, ..., X_T\)</span> and <span class="math inline">\(Y\)</span>, under some constraints (Equation <span class="math inline">\(\ref{eq:smcca2}\)</span>). In SmCCNet, the
sparsity constraint functions <span class="math inline">\(P_t(\cdot), t
= 1,2,...,T\)</span>, are the least absolute shrinkage and selection
operators (LASSO). The weighted version corresponds to <span class="math inline">\(a_{i,j}, b_{i}\)</span> (also called scaling
factors), are not all equal; the unweighted version corresponds to <span class="math inline">\(a_{i,j} = b_i = 1\)</span> for all <span class="math inline">\(i,j = 1,2,..., T\)</span>.</p>
<p><span class="math display">\[\begin{equation}
\begin{split}
\label{eq:smcca2}
&amp; (w_1, w_2,...,w_T)\!  =\! \arg\max_{\tilde w_1, \tilde w_2,...,
\tilde w_T} \!\left(\sum_{i&lt;j;i,j=1,2,...,T} \!a_{i,j}\tilde
w_i^TX_i^TX_j\tilde w_j\! +\! \sum_{i=1}^{T}b_{i}  \tilde
w_i^TX_i^TY\!\right),\\
&amp;  \hspace{.2in}\text{subject to } \lVert\tilde w_t\rVert^2 = 1,
P_t(\tilde w_t) \leq c_t, t = 1,2,...,T.           
\end{split}
\end{equation}\]</span></p>
<p>The sparsity penalties <span class="math inline">\(c_t\)</span>
influence how many features will be included in each subnetwork. With
pre-selected sparsity penalties, the SmCCNet algorithm creates a network
similarity matrix based on SmCCA canonical weights from repeated
subsampled omics data and the phenotype, and then finds multi-omics
modules that are relevant to the phenotype. The subsampling scheme
improves network robustness by analyzing a subset of omics features
multiple times and forms a final similarity matrix by aggregating
results from each subsampling step. The general workflow (Figure <span class="math inline">\(\ref{fig:flowchart}\)</span>) involves three
steps:</p>
<ul>
<li>Step I: Determine SmCCA sparsity penalties <span class="math inline">\(c_t\)</span>. The user can select the penalties
for omics feature selection based on the study purpose and/or prior
knowledge. Alternatively, one can pick sparsity penalties based on a
K-fold cross validation (CV) procedure that minimizes the total
prediction error (Figure <span class="math inline">\(\ref{fig:CVflow}\)</span>). The K-fold CV
procedure ensures selected penalties to be generalizable to similar
independent data sets and prevents over-fitting.</li>
<li>Step II: Randomly subsample omics features without replacement,
apply SmCCA with chosen penalties, and compute a feature relationship
matrix for each subset. Repeat the process many times and define the
similarity matrix to be the average of all feature relationship
matrices.</li>
<li>Step III: Apply hierarchical tree cutting to the similarity matrix
to find the multi-omics networks. This step simultaneously identifies
multiple subnetworks.</li>
</ul>
<div class="figure">
<img src="figures/SmCCNet-Quant.jpg" alt="SmCCNet workflow overview for Quantitative Phenotype. X1, X2, and X3 are three omics data types for the same set of n subjects. Y indicates a quantitative phenotype measure for those n subjects. Note that the flowchart demonstrate workflow for three omics data, it is also compatible with more than three omics data or two omics data." width="100%"><p class="caption">
SmCCNet workflow overview for Quantitative Phenotype. X1, X2, and X3 are
three omics data types for the same set of n subjects. Y indicates a
quantitative phenotype measure for those n subjects. Note that the
flowchart demonstrate workflow for three omics data, it is also
compatible with more than three omics data or two omics data.
</p>
</div>
</div>
<div class="section level3">
<h3 id="smccnet-package">SmCCNet package<a class="anchor" aria-label="anchor" href="#smccnet-package"></a>
</h3>
<p>The SmCCNet package has the following dependencies:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/psolymos/pbapply" class="external-link">pbapply</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://Matrix.R-forge.R-project.org" class="external-link">Matrix</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r.igraph.org/" class="external-link">igraph</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/KechrisLab/SmCCNet" class="external-link">SmCCNet</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/DavisVaughan/furrr" class="external-link">furrr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://future.futureverse.org" class="external-link">future</a></span><span class="op">)</span></span></code></pre></div>
<!-- The SmCCNet package consists of two R scripts: -->
<!-- ```{r source, eval = TRUE} -->
<!-- source("../R/ModifiedPMA.R") -->
<!-- source("../R/SmCCNetSource.R") -->
<!-- ``` -->
<p>The older version of the SmCCNet package includes four (external)
functions:</p>
<ul>
<li>
<strong>getRobustPseudoWeights()</strong>: Compute aggregated
(SmCCA) canonical weights.</li>
<li>
<strong>getAbar()</strong>: Calculate similarity matrix based on
canonical weights.</li>
<li>
<strong>getMultiOmicsModules()</strong>: Perform hierarchical tree
cutting on the similarity matrix and extract clades with multi-omics
features.</li>
<li>
<strong>plotMultiOmicsNetwork()</strong>: Plot (pruned or full)
multi-omics subnetworks.</li>
</ul>
<p>In the updated package, all functions except for
<strong>getAbar</strong> are retired from the package, additional
functions have been added to the package to perform single-/multi-omics
SmCCNet with quantitative/binary phenotype, and their use is illustrated
in this vignette:</p>
<ul>
<li>
<strong>aggregateCVSingle()</strong>: Saving cross-validation result
as the cross-validation table into the working directory and provide
recommendation on the penalty term selection.</li>
<li>
<strong>classifierEval()</strong>: Evaluate binary classifier’s
performance with respect to user-selected metric (accuracy, auc score,
precision, recall, f1).</li>
<li>
<strong>dataPreprocess()</strong>: A simple pipeline to preprocess
the data before running SmCCNet (center, scale, coefficient of variation
filtering and regressing out covariates).</li>
<li>
<strong>fastAutoSmCCNet()</strong>: Automated SmCCNet automatically
identifies the project problem (single-omics vs multi-omics), and type
of analysis (CCA for quantitative phenotype vs. PLS for binary
phenotype) based on the input data that is provided. This method
automatically preprocess data, choose scaling factors, subsampling
percentage, and optimal penalty terms, then runs through the complete
SmCCNet pipeline without the requirement for users to provide additional
information. This function will store all the subnetwork information to
a user-provided directory, as well as return all the global network and
evaluation information. Refer to the automated SmCCNet vignette for more
information.</li>
<li>
<strong>getCanWeightsMulti()</strong>: Run Sparse Multiple Canonical
Correlation Analysis (SmCCA) and return canonical weight.</li>
<li>
<strong>getCanCorMulti()</strong>: Get canonical correlation value
for SmCCA given canonical weight vectors and scaling factors.</li>
<li>
<strong>getRobustWeightsMulti()</strong>: SmCCNet algorithm with
multi-omics data and quantitative phenotype. Calculate the canonical
weights for SmCCA.</li>
<li>
<strong>getRobustWeightsMultiBinary()</strong>: SmCCNet algorithm
with multi-omics data and binary phenotype. First, SmCCA is used to
identify relationship between omics (exlude phenotype). Then, after
highly connected omics features are selected in step 1, SPLSDA is used
to identify relationships between these omics features and phenotype(s).
The sparse PLSDA algorithm for binary outcome first compute PLS by
assuming outcome is continuous, and extracts multiple latent factors,
then uses latent factors to fit the logistic regression, and weights
latent factor by regression parameters.</li>
<li>
<strong>getRobustWeightsSingle()</strong>: Compute aggregated
(SmCCA) canonical weights for single omics data with quantitative
phenotype.</li>
<li>
<strong>getRobustWeightsSingleBinary()</strong>: Compute aggregated
(SmCCA) canonical weights for single omics data with binary
phenotype.</li>
<li>
<strong>getOmicsModules()</strong>: Perform hierarchical tree
cutting on the similarity matrix and extract clades with omics
features.</li>
<li>
<strong>networkPruning()</strong>: Extract summarization scores (the
first 3 NetSHy/regular prinicipal components) for specified network
module with given network size. The omics features will be ranked based
on PageRank algorithm, then the top <span class="math inline">\(m\)</span> omics features (where <span class="math inline">\(m\)</span> is the specified subnetwork size) will
be included into the final subnetwork to generate the summarization
score. For the PC score, the correlation with respect to the phenotype
of interest will be calculated and stored. In addition, the correlation
between individual omics features and the detected phenotype (with the
<strong>Pheno</strong> argument) will also be recorded. The final
subnetwork adjacency matrix will be stored into the user-specified
working directory of interest.</li>
<li>
<strong>scalingFactorInput()</strong>: After inputing the annotation
of omics data, it uses prompts to ask the user to supply the scaling
factor intended for the SmCCNet algorithm to prioritize the correlation
structure of interest. All scaling factor values supplied should be
numeric and nonnegative.</li>
<li>
<strong>summarizeNetSHy()</strong>: Implement NetSHy network
summarization via a hybrid approach to summarize network by considering
the network topology with the Laplacian matrix.</li>
</ul>
<p>More details on above functions can be found in the package
manual.</p>
</div>
</div>
<div class="section level2">
<h2 id="smccnet-workflow-with-a-synthetic-dataset">SmCCNet workflow with a synthetic dataset<a class="anchor" aria-label="anchor" href="#smccnet-workflow-with-a-synthetic-dataset"></a>
</h2>
<div class="section level3">
<h3 id="synthetic-dataset">Synthetic dataset<a class="anchor" aria-label="anchor" href="#synthetic-dataset"></a>
</h3>
<p>As an example, we consider a synthetic data set with 500 genes <span class="math inline">\((X_1)\)</span> and 100 miRNAs <span class="math inline">\((X_2)\)</span> expression levels measured for 358
subjects, along with a quantitative phenotype <span class="math inline">\((Y)\)</span>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">ExampleData</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">X1</span><span class="op">[</span> , <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          Gene_1   Gene_2   Gene_3   Gene_4   Gene_5   Gene_6</span></span>
<span><span class="co">## Samp_1 22.48570 40.35372 31.02575 20.84721 26.69729 30.20545</span></span>
<span><span class="co">## Samp_2 37.05885 34.05223 33.48702 23.53146 26.75463 31.73594</span></span>
<span><span class="co">## Samp_3 20.53077 31.66962 35.18957 20.95254 25.01883 32.15723</span></span>
<span><span class="co">## Samp_4 33.18689 38.48088 18.89710 31.82330 34.04938 38.79989</span></span>
<span><span class="co">## Samp_5 28.96198 41.06049 28.49496 18.37449 30.81524 24.00454</span></span>
<span><span class="co">## Samp_6 18.05983 29.55471 32.54002 29.68452 26.19996 26.76684</span></span></code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">X2</span><span class="op">[</span> , <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##           Mir_1    Mir_2    Mir_3    Mir_4    Mir_5     Mir_6</span></span>
<span><span class="co">## Samp_1 15.22391 17.54583 15.78472 14.89198 10.34821  9.689755</span></span>
<span><span class="co">## Samp_2 16.30697 16.67283 13.36153 14.48855 12.66090 11.333613</span></span>
<span><span class="co">## Samp_3 16.54512 16.73501 14.61747 17.84527 13.82279 11.329333</span></span>
<span><span class="co">## Samp_4 13.98690 16.20743 16.29308 17.72529 12.30056  9.844108</span></span>
<span><span class="co">## Samp_5 16.33833 17.39387 16.39792 15.85373 13.38767 10.599414</span></span>
<span><span class="co">## Samp_6 14.54110 16.51999 14.73958 15.87504 13.21359 10.922393</span></span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##           Pheno</span></span>
<span><span class="co">## Samp_1 235.0674</span></span>
<span><span class="co">## Samp_2 253.5450</span></span>
<span><span class="co">## Samp_3 234.2050</span></span>
<span><span class="co">## Samp_4 281.0354</span></span>
<span><span class="co">## Samp_5 245.4478</span></span>
<span><span class="co">## Samp_6 189.6231</span></span></code></pre>
<p>Denote the number of features in <span class="math inline">\(X_1
\&amp; X_2\)</span> as <span class="math inline">\(p_1 \&amp;
p_2\)</span> respectively, and the number of subjects as <span class="math inline">\(n\)</span>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">X1</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">X2</span><span class="op">)</span></span>
<span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">X1</span><span class="op">)</span></span>
<span><span class="va">AbarLabel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">X1</span>, <span class="va">X2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Although SmCCNet does not require normality, it calculates the
Pearson correlation between linear combinations of omics features and
the phenotype, which assumes finite variances and finite covariance. It
is necessary to include a transformation if the data are skewed. The
algorithm also requires the data to be standardizable (i.e. none of the
data matrices include a column with zero variance.)</p>
</div>
<div class="section level3">
<h3 id="step-i--preprocessing">Step I. Preprocessing<a class="anchor" aria-label="anchor" href="#step-i--preprocessing"></a>
</h3>
<p>The first step is to preprocess the single omics data. All canonical
correlation analysis (CCA) methods require data sets to be standardized
(centered and scaled) by columns (e.g. features) to ensure the
equivalent contribution of each feature when maximizing covariance. In
addition, there are some other optional steps to preprocess the data.
Note that when employing both standardization and covariate adjustment,
it is essential to apply standardization prior to covariate adjustment.
This sequencing ensures our best efforts to meet the assumptions of
linear regression. In our data preprocessing pipeline, the following
options are given:</p>
<ul>
<li>
<strong>CoV Filtering</strong>: Filter features based on
coefficients of variation (CoV).</li>
<li>
<strong>Standardization</strong>: Center and/or scale data.</li>
<li>
<strong>Adjust for Covariates</strong>: Regress out specified
covariates and return residuals</li>
</ul>
<p>Below is an example of data preprocessing with only feature filtering
and standardization, in this case, there is no covariate adjustment, the
coeffcient of variation filtering quantile is 0.2 (meaning that features
with CoV smaller than 20% quantile of CoV will be filtered out), and
data are centered and scaled:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># define data list</span></span>
<span><span class="va">X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">X1</span>, <span class="va">X2</span><span class="op">)</span></span>
<span><span class="co"># preprocess data</span></span>
<span><span class="va">processed_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">X</span>, <span class="kw">function</span><span class="op">(</span><span class="va">Xmatrix</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="../reference/dataPreprocess.html">dataPreprocess</a></span><span class="op">(</span>X <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">Xmatrix</span><span class="op">)</span>, covariates <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>                  is_cv <span class="op">=</span> <span class="cn">TRUE</span>, cv_quantile <span class="op">=</span> <span class="fl">0.2</span>, center <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                 scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></span>
<span><span class="co"># re-standardize -omics data if regress-out approach is used (covariates != NULL)</span></span>
<span><span class="va">processed_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">processed_data</span>, <span class="va">scale</span><span class="op">)</span></span>
<span><span class="co"># if preprocess feature is used, X need to be overrided with the following code </span></span>
<span><span class="va">X</span> <span class="op">&lt;-</span> <span class="va">processed_data</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="step-ii-determine-optimal-sparsity-penalties-through-cross-validation-optional">Step II: Determine optimal sparsity penalties through
cross-validation (optional)<a class="anchor" aria-label="anchor" href="#step-ii-determine-optimal-sparsity-penalties-through-cross-validation-optional"></a>
</h3>
<p>For two omics data, to find the optimal sparsity penalties <span class="math inline">\(c_1, c_2\)</span>, we apply a K-fold CV on the
synthetic data (Figure <span class="math inline">\(\ref{fig:CVflow}\)</span>). Note that under LASSO
constraints, <span class="math inline">\(1\leq c_1\leq\sqrt{p_1s_1},
1\leq c_2\leq\sqrt{p_2s_2}\)</span>, where <span class="math inline">\(p_1,p_2\)</span> denote the number of features in
omics data <span class="math inline">\(X_1, X_2\)</span> respectively,
and <span class="math inline">\(s_1, s_2\)</span> are the proportions of
<span class="math inline">\(X_1, X_2\)</span> features to be sampled
every time. The sparse penalties <span class="math inline">\(c_1,
c_2\)</span> can be re-parametrized as <span class="math inline">\(0&lt;l_1, l_2\leq1\)</span>, such that <span class="math inline">\(c_1 = \max\{1, l_1\sqrt{p_1s_1}\}, c_2 = \max\{1,
l_2\sqrt{p_2s_2}\}\)</span>. Large penalty values correspond to more
features in each subnetwork, while small penalties correspond to fewer
features. Here is the list of parameters that need to be specified to
perform a grid search for the penalty parameters:</p>
<ul>
<li>
<span class="math inline">\(K\)</span>: Number of folds in CV.
Typically a 5-fold CV is sufficient. If the training set contains too
few (e.g. <span class="math inline">\(&lt;30\)</span>) samples, or the
test or training set becomes unscalable, then choose a smaller <span class="math inline">\(K\)</span>.</li>
<li>
<span class="math inline">\(CCcoef:\)</span> Coefficients, <span class="math inline">\((a_{i,j}, b_{i})\)</span> in Equation <span class="math inline">\(\ref{eq:smcca2}\)</span>, for the weighted SmCCA.
It can be either supplied manually or interatively (see code
below).</li>
<li>
<span class="math inline">\(PenExpand\)</span>: A penalty option
matrix for <span class="math inline">\(X_1, X_2\)</span>. Each row of
<span class="math inline">\(PenExpand\)</span> is a pair of penalty
options <span class="math inline">\(l = (l_1, l_2)\)</span>, where <span class="math inline">\(0&lt;l_1, l_2&lt;1\)</span>. Larger penalties
correspond to more features to be included in each subnetwork.
Typically, it is not necessary to search the entire range (0, 1), and a
smaller grid (e.g. <span class="math inline">\(0&lt;l_1,
l_2&lt;.5\)</span>) may be sufficient.</li>
<li>
<span class="math inline">\(X\)</span> A list of omics data, where
each omics data should contain same set and order of subjects.</li>
<li>
<span class="math inline">\(num\_omics\)</span> The total number of
omics data in <span class="math inline">\(X\)</span>.</li>
<li>
<span class="math inline">\(tuneLength\)</span> The total number of
candidate penalty terms for each penalty parameter.</li>
<li>
<span class="math inline">\(minTune\)</span> The minimumally
possible value for penalty parameters.</li>
<li>
<span class="math inline">\(maxTune\)</span> The maximally possible
tuning value for penalty parameters.</li>
</ul>
<div class="figure">
<img src="figures/SmCCNetCV.png" alt="SmCCNet K-fold CV. The best penalty pairs are chosen based on the smallest total prediction error." width="100%"><p class="caption">
SmCCNet K-fold CV. The best penalty pairs are chosen based on the
smallest total prediction error.
</p>
</div>
<p>In all the code chunks below, if a user plans to use more than 2
omics data, please change the code accordingly where a <span class="math inline">\(**\)</span> is marked on the comment line:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># number of folds in K-fold CV.</span></span>
<span><span class="va">K</span> <span class="op">&lt;-</span> <span class="fl">3</span> </span>
<span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">X1</span><span class="op">)</span></span>
<span><span class="co"># create a list of omics data **</span></span>
<span><span class="va">X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">X1</span>, <span class="va">X2</span><span class="op">)</span></span>
<span><span class="co"># number of omics dataset **</span></span>
<span><span class="va">num_omics</span> <span class="op">&lt;-</span> <span class="fl">2</span></span>
<span><span class="co"># tuning parameter candidate length for each omics data</span></span>
<span><span class="va">tuneLength</span> <span class="op">&lt;-</span> <span class="fl">5</span></span>
<span><span class="co"># tuning parameter candadate range for each omics data</span></span>
<span><span class="va">minTune</span> <span class="op">&lt;-</span> <span class="fl">0.1</span></span>
<span><span class="va">maxTune</span> <span class="op">&lt;-</span> <span class="fl">0.5</span></span>
<span><span class="co"># create empty matrix to store all possible penalty parameters</span></span>
<span><span class="va">penSelect</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">0</span>, nrow <span class="op">=</span> <span class="va">tuneLength</span>, ncol <span class="op">=</span> <span class="va">num_omics</span><span class="op">)</span></span>
<span><span class="co"># create sparsity penalty options.</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">Idx</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">penSelect</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">{</span></span>
<span>  <span class="va">penSelect</span><span class="op">[</span>,<span class="va">Idx</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span>from <span class="op">=</span> <span class="va">minTune</span>, </span>
<span>                         to <span class="op">=</span> <span class="va">maxTune</span>, </span>
<span>                         length.out <span class="op">=</span> <span class="va">tuneLength</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co"># expand grid</span></span>
<span><span class="co"># convert matrix to list of columns</span></span>
<span><span class="va">list_cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">as.list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">penSelect</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># generate all possible combinations</span></span>
<span><span class="va">PenExpand</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">expand.grid</span>, <span class="va">list_cols</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># set a CV directory.</span></span>
<span><span class="va">CVDir</span> <span class="op">&lt;-</span> <span class="st">"Example3foldCV/"</span> </span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">CVDir</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="create-test-and-training-data-sets-">Create test and training data sets.<a class="anchor" aria-label="anchor" href="#create-test-and-training-data-sets-"></a>
</h4>
<p>First, we need to split the data (<span class="math inline">\(X_1,
X_2, Y\)</span>) into test and training sets (Figure <span class="math inline">\(\ref{fig:CVflow}\)</span>, Step I.1). All CCA
methods require data sets to be standardized (centered and scaled) by
columns (e.g. features). We have included the standardization step
within the SmCCNet algorithm. However, for the CV procedure, we
recommend to standardize the training and test sets upfront, since this
helps to choose the number of CV folds <span class="math inline">\(K\)</span>. If any data set can not be
standardized, we recommend to reduce <span class="math inline">\(K\)</span>. In the code below, we show how to
create CV data sets. The standardized training and test data sets will
be saved under the “Example3foldCV/” directory.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">12345</span><span class="op">)</span> <span class="co"># set random seed.</span></span>
<span></span>
<span><span class="co"># split data into folds</span></span>
<span><span class="va">foldIdx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">X</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">X</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, <span class="va">K</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">folddata</span> <span class="op">&lt;-</span> <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">foldIdx</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">Y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span></span>
<span>    <span class="va">X_train</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="va">X_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="va">Y_train</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="va">Y_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">{</span></span>
<span>      <span class="va">X_train</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/scale.html" class="external-link">scale</a></span><span class="op">(</span><span class="va">X</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="op">-</span><span class="va">foldIdx</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span>,<span class="op">]</span><span class="op">)</span></span>
<span>      <span class="va">X_test</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/scale.html" class="external-link">scale</a></span><span class="op">(</span><span class="va">X</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">foldIdx</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span>,<span class="op">]</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>    <span class="va">Y_train</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/scale.html" class="external-link">scale</a></span><span class="op">(</span><span class="va">Y</span><span class="op">[</span><span class="op">-</span><span class="va">foldIdx</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span>,<span class="op">]</span><span class="op">)</span></span>
<span>    <span class="va">Y_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/scale.html" class="external-link">scale</a></span><span class="op">(</span><span class="va">Y</span><span class="op">[</span><span class="va">foldIdx</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span>,<span class="op">]</span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>X_train <span class="op">=</span> <span class="va">X_train</span>, X_test <span class="op">=</span> <span class="va">X_test</span>,Y_train <span class="op">=</span> <span class="va">Y_train</span>, </span>
<span>                Y_test <span class="op">=</span> <span class="va">Y_test</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="co"># name each fold of data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">folddata</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'Fold_'</span>, <span class="fl">1</span><span class="op">:</span><span class="va">K</span><span class="op">)</span></span>
<span><span class="co"># saving all preliminary data into local directory for reproducibility purpose</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span><span class="va">folddata</span>, <span class="va">PenExpand</span>, </span>
<span>       file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">CVDir</span>, <span class="st">"CVData.RData"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="set-scaling-factors">Set Scaling Factors<a class="anchor" aria-label="anchor" href="#set-scaling-factors"></a>
</h4>
<p>It is important to set scaling factors <span class="math inline">\(a_{i,j}\)</span>’s and <span class="math inline">\(b_{i}\)</span>’s for each pairwise correlation to
prioritize correlation structure(s) of interest. Below is the function
that prompts the user to define the scaling factor intended for the
analysis: default is (1,1,1) for two omics if the first line of code
below is run. If the user prefers to set up scaling factors in different
ways such as (1,10,10) where 10 is set for omics-phenotype correlation
to emphasize the omics-phenotype correlation, the user can use the
interactive function below. This is especially effective when more than
two omics data are used since the order of pairwise combinations can be
extremely complicated. Note that the argument (DataType) for interactive
function below should strictly follow the order of the data list used
for SmCCA computation, followed by ‘phenotype’. For instance, in the
example data, <span class="math inline">\(X_1\)</span> stands for mRNA,
<span class="math inline">\(X_2\)</span> stands for miRNAs, then the
order of the argument is set to (‘mRNA’, ‘miRNA’, ‘phenotype’). In the
example below, we use <span class="math inline">\((1,1,1)\)</span> as
the scaling factor to demonstrate the result.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># default</span></span>
<span><span class="va">scalingFactor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/combn.html" class="external-link">combn</a></span><span class="op">(</span><span class="va">num_omics</span> <span class="op">+</span> <span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># interactive **</span></span>
<span><span class="va">scalingFactor</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scalingFactorInput.html">scalingFactorInput</a></span><span class="op">(</span>DataType <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'mRNA'</span>, <span class="st">'miRNA'</span>, <span class="st">'phenotype'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="run-k-fold-cv">Run K-fold CV<a class="anchor" aria-label="anchor" href="#run-k-fold-cv"></a>
</h4>
<p>For each of the K-folds, we compute the prediction error for each
penalty pair option (Figure <span class="math inline">\(\ref{fig:CVflow}\)</span>, Step I.2). Since there
is no subsampling step for cross-validation, we run through
cross-validation with nested for loop. However, if the omics data are
extremely high-dimensional, we recommend using the R package
<strong>parallel</strong> to parallelize the for loop, or use
<strong>fastAutoSmCCNet()</strong> directly.
<strong>fastAutoSmCCNet()</strong> is the package built-in function that
streamline the pipeline with single line of code, and the
cross-validation step is parallelized with <strong>future_map()</strong>
in the <strong>furrr</strong> package.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load cross-validation data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">CVDir</span>, <span class="st">"CVData.RData"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># create an empty list for storing CV result for each fold</span></span>
<span><span class="va">CVResult</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">CVidx</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">K</span><span class="op">)</span></span>
<span><span class="op">{</span>  </span>
<span>    <span class="co"># set scaling factor</span></span>
<span>    <span class="va">CCcoef</span> <span class="op">&lt;-</span> <span class="va">scalingFactor</span></span>
<span>    <span class="co"># create empty vector for storing cross-validation result</span></span>
<span>    <span class="va">RhoTrain</span> <span class="op">&lt;-</span> <span class="va">RhoTest</span> <span class="op">&lt;-</span> <span class="va">DeltaCor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">PenExpand</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="kw">for</span><span class="op">(</span><span class="va">idx</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">PenExpand</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">{</span></span>
<span>      <span class="co"># consider one pair of sparsity penalties at a time.</span></span>
<span>      <span class="va">l</span> <span class="op">&lt;-</span> <span class="va">PenExpand</span><span class="op">[</span><span class="va">idx</span>, <span class="op">]</span></span>
<span>      <span class="co"># run SmCCA on the subsamples (Figure 1, Step II)</span></span>
<span>      <span class="va">Ws</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getCanWeightsMulti.html">getCanWeightsMulti</a></span><span class="op">(</span><span class="va">folddata</span><span class="op">[[</span><span class="va">CVidx</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"X_train"</span><span class="op">]</span><span class="op">]</span>, </span>
<span>                           Trait <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">folddata</span><span class="op">[[</span><span class="va">CVidx</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"Y_train"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, </span>
<span>                           Lambda <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">l</span><span class="op">)</span>, NoTrait <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                           CCcoef <span class="op">=</span> <span class="va">CCcoef</span><span class="op">)</span></span>
<span>      <span class="co"># evaluate the canonical correlation for training and testing data</span></span>
<span>      <span class="va">rho.train</span> <span class="op">&lt;-</span>  <span class="fu"><a href="../reference/getCanCorMulti.html">getCanCorMulti</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">folddata</span><span class="op">[[</span><span class="va">CVidx</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"X_train"</span><span class="op">]</span><span class="op">]</span>, </span>
<span>                                Y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">folddata</span><span class="op">[[</span><span class="va">CVidx</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"Y_train"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, </span>
<span>                                CCWeight <span class="op">=</span> <span class="va">Ws</span>, </span>
<span>                                CCcoef <span class="op">=</span> <span class="va">CCcoef</span><span class="op">)</span></span>
<span>      <span class="va">rho.test</span> <span class="op">&lt;-</span>  <span class="fu"><a href="../reference/getCanCorMulti.html">getCanCorMulti</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">folddata</span><span class="op">[[</span><span class="va">CVidx</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"X_test"</span><span class="op">]</span><span class="op">]</span>, </span>
<span>                                Y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">folddata</span><span class="op">[[</span><span class="va">CVidx</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"Y_test"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, </span>
<span>                                CCWeight <span class="op">=</span> <span class="va">Ws</span>, </span>
<span>                                CCcoef <span class="op">=</span> <span class="va">CCcoef</span><span class="op">)</span></span>
<span></span>
<span>      <span class="co"># store cv result  </span></span>
<span>      <span class="va">RhoTrain</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">rho.train</span>, digits <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span>      <span class="va">RhoTest</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">rho.test</span>, digits <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span>      <span class="va">DeltaCor</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">rho.train</span> <span class="op">-</span> <span class="va">rho.test</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="op">}</span></span>
<span>    </span>
<span>    <span class="co"># record prediction errors for given CV fold and all sparsity penalty </span></span>
<span>    <span class="co"># options.</span></span>
<span>    <span class="va">CVResult</span><span class="op">[[</span><span class="va">CVidx</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">RhoTrain</span>, <span class="va">RhoTest</span>, <span class="va">DeltaCor</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="extract-penalty-pair-with-the-smallest-total-prediction-error">Extract penalty pair with the smallest total prediction error<a class="anchor" aria-label="anchor" href="#extract-penalty-pair-with-the-smallest-total-prediction-error"></a>
</h4>
<p>Finally, we extract the total prediction errors (Figure <span class="math inline">\(\ref{fig:CVflow}\)</span>, Step I.3) and conclude
the best penalty pair from the grid search as the pair with the smallest
error (Figure <span class="math inline">\(\ref{fig:CVflow}\)</span>,
Step I.4). There are different ways to select the best penalty terms,
one of the simplest way is to minimize discrepancy between the training
canonical correlation and the testing canonical correlation. However,
this method does not take the magnitude of testing canonical correlation
into account, which means it may end up selecting the penalty term with
smaller canonical correlation (low signal). For instance, if a certain
penalty term yields the training canonical correlation of 0.7, with the
testing canonical correlation of 0.4, and another penalty term yield the
training canonical correlation of 0.4, with the testing canonical
correlation of 0.2, minimizing training and testing canonical
correlation selects the latter. Therefore, in this step, we want to
minimized the scaled prediction error, which is defined as:</p>
<p><span class="math display">\[\begin{equation}
scaledPredErr = \frac{|trainCC - testCC|}{|testCC|},
\end{equation}\]</span></p>
<p>where <span class="math inline">\(trainCC\)</span> and <span class="math inline">\(testCC\)</span> is defined as the training
canonical correlation and testing canonical correlation respectively.
Below is the aggregated cross-validation evaluation result:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># aggregate CV result and select the best penalty term</span></span>
<span><span class="va">AggregatedCVResult</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/funprog.html" class="external-link">Reduce</a></span><span class="op">(</span><span class="st">"+"</span>, <span class="va">CVResult</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">CVResult</span><span class="op">)</span></span>
<span><span class="co"># calculate the evaluation metric of interest</span></span>
<span><span class="va">EvalMetric</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">AggregatedCVResult</span>, <span class="fl">1</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span><span class="va">x</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">x</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></span>
<span><span class="co"># determine the best CV result</span></span>
<span><span class="va">optIdx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.min.html" class="external-link">which.min</a></span><span class="op">(</span><span class="va">EvalMetric</span><span class="op">)</span></span></code></pre></div>
<p>We can visualize the scaled prediction errors with a contour plot
(Figure <span class="math inline">\(\ref{fig:contourPlot}\)</span>),
note that this method only works for two omics data, if more than two
omics data are used, please construct contour plot for each pair of
omics data respectively.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://plotly-r.com" class="external-link">plotly</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/hadley/reshape" class="external-link">reshape2</a></span><span class="op">)</span></span>
<span><span class="va">f1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  family <span class="op">=</span> <span class="st">"Arial, sans-serif"</span>,</span>
<span>  size <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  color <span class="op">=</span> <span class="st">"black"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">f2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  family <span class="op">=</span> <span class="st">"Old Standard TT, serif"</span>,</span>
<span>  size <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  color <span class="op">=</span> <span class="st">"black"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">a</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  title <span class="op">=</span> <span class="st">"l1"</span>,</span>
<span>  titlefont <span class="op">=</span> <span class="va">f1</span>,</span>
<span>  showticklabels <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  tickfont <span class="op">=</span> <span class="va">f2</span></span>
<span><span class="op">)</span></span>
<span><span class="va">b</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  title <span class="op">=</span> <span class="st">"l2"</span>,</span>
<span>  titlefont <span class="op">=</span> <span class="va">f1</span>,</span>
<span>  showticklabels <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  tickfont <span class="op">=</span> <span class="va">f2</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># create melt data</span></span>
<span><span class="va">PenExpandMelt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">PenExpand</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">]</span>, <span class="va">EvalMetric</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">PenExpandMelt</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'l1'</span>, <span class="st">'l2'</span><span class="op">)</span></span>
<span><span class="va">hmelt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/reshape2/man/melt.html" class="external-link">melt</a></span><span class="op">(</span><span class="va">PenExpandMelt</span>, id.vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"l1"</span>, <span class="st">"l2"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">contourPlot</span> <span class="op">&lt;-</span> <span class="fu">plot_ly</span><span class="op">(</span><span class="va">hmelt</span>, x <span class="op">=</span> <span class="op">~</span><span class="va">l1</span>, y <span class="op">=</span> <span class="op">~</span><span class="va">l2</span>, z <span class="op">=</span> <span class="op">~</span><span class="va">value</span>, </span>
<span>                       type <span class="op">=</span> <span class="st">"contour"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/graphics/layout.html" class="external-link">layout</a></span><span class="op">(</span>xaxis <span class="op">=</span> <span class="va">a</span>, yaxis <span class="op">=</span> <span class="va">b</span>, showlegend <span class="op">=</span> <span class="cn">TRUE</span>, legend <span class="op">=</span> <span class="va">f1</span><span class="op">)</span>  </span>
<span><span class="co"># orca preinstalltion is required for next step: </span></span>
<span><span class="co"># https://github.com/plotly/orca#installation</span></span>
<span><span class="va">contourPlot</span></span></code></pre></div>
<div class="figure">
<img src="figures/TotalPredictionError.jpeg" alt="Total scaled prediction error contour plot for evaluation metric (defined as prediction error/test cc). The x- and y-axes indicate LASSO penalties considered for mRNA and miRNA, respectively. Blue to yellow scale indicates increasing total scaled prediction error." width="100%"><p class="caption">
Total scaled prediction error contour plot for evaluation metric
(defined as prediction error/test cc). The x- and y-axes indicate LASSO
penalties considered for mRNA and miRNA, respectively. Blue to yellow
scale indicates increasing total scaled prediction error.
</p>
</div>
<p>For the synthetic data set, the optimal penalty pair that gives the
smallest prediction error is <span class="math inline">\((l_1, l_2) =
(0.1, 0.1)\)</span>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># combine CV evaluation result with penalty candidates </span></span>
<span><span class="va">overallCVInfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">PenExpand</span>, <span class="va">AggregatedCVResult</span>, scaledPredError <span class="op">=</span> <span class="va">EvalMetric</span><span class="op">)</span></span>
<span><span class="co"># set column names for penalty terms </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">overallCVInfo</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">num_omics</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'l'</span>,<span class="fl">1</span><span class="op">:</span><span class="va">num_omics</span><span class="op">)</span></span>
<span><span class="co"># save overall CV result </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.csv</a></span><span class="op">(</span><span class="va">overallCVInfo</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">CVDir</span>, <span class="st">'overallCVInfo.csv'</span><span class="op">)</span>, </span>
<span>          row.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co"># print out the best CV penalty pair and associated result</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">overallCVInfo</span><span class="op">[</span><span class="va">optIdx</span>,<span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>Table <span class="math inline">\(\ref{tab:errorTable}\)</span> shows
the total prediction error (scaledPredError) for all penalty options.
Note that in this example, we are only including 25 optional penalty
pairs. The fourth column (RhoTest) records the aggregated canonical
correlations for the test data set.</p>
<table class="table">
<caption>Total Prediction Error from a 3-fold CV for the synthetic
dataset</caption>
<thead><tr class="header">
<th align="right">l1</th>
<th align="right">l2</th>
<th align="right">RhoTrain</th>
<th align="right">RhoTest</th>
<th align="right">DeltaCor</th>
<th align="right">scaledPredError</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">0.1</td>
<td align="right">0.1</td>
<td align="right">2.143180</td>
<td align="right">2.021583</td>
<td align="right">0.3073917</td>
<td align="right">0.1520549</td>
</tr>
<tr class="even">
<td align="right">0.2</td>
<td align="right">0.1</td>
<td align="right">2.298900</td>
<td align="right">1.937143</td>
<td align="right">0.3617541</td>
<td align="right">0.1867462</td>
</tr>
<tr class="odd">
<td align="right">0.3</td>
<td align="right">0.1</td>
<td align="right">2.371103</td>
<td align="right">1.802180</td>
<td align="right">0.5689212</td>
<td align="right">0.3156850</td>
</tr>
<tr class="even">
<td align="right">0.4</td>
<td align="right">0.1</td>
<td align="right">2.226837</td>
<td align="right">1.281533</td>
<td align="right">0.9453013</td>
<td align="right">0.7376331</td>
</tr>
<tr class="odd">
<td align="right">0.5</td>
<td align="right">0.1</td>
<td align="right">2.257677</td>
<td align="right">1.161217</td>
<td align="right">1.0964606</td>
<td align="right">0.9442343</td>
</tr>
<tr class="even">
<td align="right">0.1</td>
<td align="right">0.2</td>
<td align="right">2.439183</td>
<td align="right">1.903217</td>
<td align="right">0.5359673</td>
<td align="right">0.2816113</td>
</tr>
<tr class="odd">
<td align="right">0.2</td>
<td align="right">0.2</td>
<td align="right">2.553920</td>
<td align="right">1.827470</td>
<td align="right">0.7264469</td>
<td align="right">0.3975151</td>
</tr>
<tr class="even">
<td align="right">0.3</td>
<td align="right">0.2</td>
<td align="right">2.606787</td>
<td align="right">1.708473</td>
<td align="right">0.8983147</td>
<td align="right">0.5257997</td>
</tr>
<tr class="odd">
<td align="right">0.4</td>
<td align="right">0.2</td>
<td align="right">2.641467</td>
<td align="right">1.553353</td>
<td align="right">1.0881148</td>
<td align="right">0.7004941</td>
</tr>
<tr class="even">
<td align="right">0.5</td>
<td align="right">0.2</td>
<td align="right">2.662887</td>
<td align="right">1.420033</td>
<td align="right">1.2428491</td>
<td align="right">0.8752253</td>
</tr>
<tr class="odd">
<td align="right">0.1</td>
<td align="right">0.3</td>
<td align="right">2.603863</td>
<td align="right">1.770733</td>
<td align="right">0.8331298</td>
<td align="right">0.4704999</td>
</tr>
<tr class="even">
<td align="right">0.2</td>
<td align="right">0.3</td>
<td align="right">2.699880</td>
<td align="right">1.663637</td>
<td align="right">1.0362454</td>
<td align="right">0.6228796</td>
</tr>
<tr class="odd">
<td align="right">0.3</td>
<td align="right">0.3</td>
<td align="right">2.738233</td>
<td align="right">1.560693</td>
<td align="right">1.1775418</td>
<td align="right">0.7544992</td>
</tr>
<tr class="even">
<td align="right">0.4</td>
<td align="right">0.3</td>
<td align="right">2.760453</td>
<td align="right">1.420607</td>
<td align="right">1.3398453</td>
<td align="right">0.9431501</td>
</tr>
<tr class="odd">
<td align="right">0.5</td>
<td align="right">0.3</td>
<td align="right">2.773497</td>
<td align="right">1.303357</td>
<td align="right">1.4701370</td>
<td align="right">1.1279621</td>
</tr>
<tr class="even">
<td align="right">0.1</td>
<td align="right">0.4</td>
<td align="right">2.703797</td>
<td align="right">1.665173</td>
<td align="right">1.0386261</td>
<td align="right">0.6237345</td>
</tr>
<tr class="odd">
<td align="right">0.2</td>
<td align="right">0.4</td>
<td align="right">2.787583</td>
<td align="right">1.537180</td>
<td align="right">1.2504062</td>
<td align="right">0.8134416</td>
</tr>
<tr class="even">
<td align="right">0.3</td>
<td align="right">0.4</td>
<td align="right">2.813273</td>
<td align="right">1.427817</td>
<td align="right">1.3854583</td>
<td align="right">0.9703334</td>
</tr>
<tr class="odd">
<td align="right">0.4</td>
<td align="right">0.4</td>
<td align="right">2.827043</td>
<td align="right">1.284917</td>
<td align="right">1.5421291</td>
<td align="right">1.2001783</td>
</tr>
<tr class="even">
<td align="right">0.5</td>
<td align="right">0.4</td>
<td align="right">2.831297</td>
<td align="right">1.185093</td>
<td align="right">1.6461995</td>
<td align="right">1.3890885</td>
</tr>
<tr class="odd">
<td align="right">0.1</td>
<td align="right">0.5</td>
<td align="right">2.763300</td>
<td align="right">1.581297</td>
<td align="right">1.1820000</td>
<td align="right">0.7474878</td>
</tr>
<tr class="even">
<td align="right">0.2</td>
<td align="right">0.5</td>
<td align="right">2.845553</td>
<td align="right">1.459457</td>
<td align="right">1.3860930</td>
<td align="right">0.9497322</td>
</tr>
<tr class="odd">
<td align="right">0.3</td>
<td align="right">0.5</td>
<td align="right">2.866427</td>
<td align="right">1.346777</td>
<td align="right">1.5196518</td>
<td align="right">1.1283621</td>
</tr>
<tr class="even">
<td align="right">0.4</td>
<td align="right">0.5</td>
<td align="right">2.872273</td>
<td align="right">1.213000</td>
<td align="right">1.6592727</td>
<td align="right">1.3679082</td>
</tr>
<tr class="odd">
<td align="right">0.5</td>
<td align="right">0.5</td>
<td align="right">2.870893</td>
<td align="right">1.121857</td>
<td align="right">1.7490378</td>
<td align="right">1.5590564</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section level3">
<h3 id="step-iii-run-smcca-with-pre-selected-penalty-term">Step III: Run SmCCA with pre-selected penalty term<a class="anchor" aria-label="anchor" href="#step-iii-run-smcca-with-pre-selected-penalty-term"></a>
</h3>
<p>With a pre-selected penalty pair, we apply SmCCA to subsampled
features of <span class="math inline">\(X_1, X_2\)</span> and <span class="math inline">\(Y\)</span>, and repeat the process to generate a
robust similarity matrix (Figure <span class="math inline">\(\ref{fig:flowchart}\)</span>, Step II). As for the
number of subsamples, a larger number of subsamples leads to more
accurate results, while a smaller number of subsamples is faster
computationally. We use 50 in this example. In general, we recommend to
subsample 500-1000 times or more. Below is the setup and description of
the subsampling parameters:</p>
<ul>
<li>
<span class="math inline">\(s = (s_1,s_2)\)</span>: Proportions of
feature subsampling from <span class="math inline">\(X_1, X_2\)</span>.
Default values are <span class="math inline">\(s_1 = 0.7, s_2 =
0.9.\)</span>
</li>
<li>
<span class="math inline">\(SubsamplingNum\)</span>: Number of
subsamples.</li>
</ul>
<p>After obtaining the canonical weight <span class="math inline">\(Ws\)</span>, which has the dimension of <span class="math inline">\(p_1 + p_2\)</span> (total number of features) by
<span class="math inline">\(p_1 s_1 + p_2 s_2\)</span> (number of
feature subsamples), the next step is to obtain the adjacency matrix by
taking the outer product of each <span class="math inline">\(W\)</span>
with itself to obtain an adjacency matrix and average the matrices to
obtain <span class="math inline">\(Abar\)</span>, a sparse matrix object
(illustrate in the next section).</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># feature sampling proportions, 0.9 for miRNA since it has less features. **</span></span>
<span><span class="va">s</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.7</span>,<span class="fl">0.9</span><span class="op">)</span> </span>
<span><span class="co"># number of subsamples.</span></span>
<span><span class="va">SubsamplingNum</span> <span class="op">&lt;-</span> <span class="fl">50</span></span>
<span><span class="co"># run SmCCA on the subsamples (Figure 1, Step II)</span></span>
<span><span class="va">Ws</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getRobustWeightsMulti.html">getRobustWeightsMulti</a></span><span class="op">(</span><span class="va">X</span>, </span>
<span>                        Trait <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span>, </span>
<span>                        NoTrait <span class="op">=</span> <span class="cn">FALSE</span>,CCcoef <span class="op">=</span> <span class="va">scalingFactor</span>,</span>
<span>                        Lambda <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">overallCVInfo</span><span class="op">[</span><span class="va">optIdx</span>,<span class="fl">1</span><span class="op">:</span><span class="va">num_omics</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                        s <span class="op">=</span> <span class="va">s</span>, </span>
<span>                        SubsamplingNum <span class="op">=</span> <span class="va">SubsamplingNum</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="step-iv-obtain-multi-omics-modules-through-network-clustering">Step IV: Obtain multi-omics modules through network clustering<a class="anchor" aria-label="anchor" href="#step-iv-obtain-multi-omics-modules-through-network-clustering"></a>
</h3>
<p>From the canonical weights obtained in the last step, we can get
multi-omics modules by constructing global adjacency matrix and applying
hierarchical tree cutting and plotting the reconstructed networks
(Figure <span class="math inline">\(\ref{fig:flowchart}\)</span>). The
edge signs are recovered from pairwise feature correlations when
visualizing the network.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># construct adjacency matrix</span></span>
<span><span class="va">Abar</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getAbar.html">getAbar</a></span><span class="op">(</span><span class="va">Ws</span>, FeatureLabel <span class="op">=</span> <span class="va">AbarLabel</span><span class="op">)</span></span>
<span><span class="co"># perform clustering based on the adjacency matrix Abar </span></span>
<span><span class="va">OmicsModule</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getOmicsModules.html">getOmicsModules</a></span><span class="op">(</span><span class="va">Abar</span>, PlotTree <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span><span class="va">Ws</span>, <span class="va">Abar</span>, <span class="va">OmicsModule</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">CVDir</span>, <span class="st">"SmCCNetWeights.RData"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="step-v-obtain-network-summarization-score-and-pruned-subnetworks">Step V: Obtain network summarization score and pruned
subnetworks<a class="anchor" aria-label="anchor" href="#step-v-obtain-network-summarization-score-and-pruned-subnetworks"></a>
</h3>
<p>The next step is to prune the network so that the unnecessary
features (nodes) will be filtered from the original network module. This
principle is based on a subject-level score of interest known as the
network summarization score. There are two different network
summarization methods: PCA and NetSHy (network summarization via a
hybrid approach, Vu et al 2023 Bioinformatics), which are specified by
the argument ‘method’. We evaluate two criteria stepwise 1)
summarization score correlation with respect to the phenotype, which is
used to verify if the summarization score for the current subnetwork has
a strong signal with respect to the phenotype and 2) The correlation
between the summarization of the current subnetwork and that of the
baseline network with a pre-defined baseline network size. This is used
to check if the addition of more molecular features introduces noise.
The stepwise approach for network pruning is:</p>
<p>In the following example, the summarization used is the first 3
NetSHy principal components. After clustering, there are 6 network
modules, with only 1 network module having more than 10 nodes. The
optimal network size after network pruning is 18. The output from this
steps contains a network adjacency matrix, summarization scores (first 3
NetSHy PCs), PC loadings and more, which are stored in a .Rdata file in
the user specified location.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># make sure there are no duplicated labels</span></span>
<span><span class="va">AbarLabel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/make.unique.html" class="external-link">make.unique</a></span><span class="op">(</span><span class="va">AbarLabel</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># create concatenate omics data for network pruning</span></span>
<span><span class="va">X_big</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">X1</span>,<span class="va">X2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># calculate feature correlation matrix</span></span>
<span><span class="va">bigCor2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">X_big</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># data type</span></span>
<span><span class="va">types</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">'gene'</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">X1</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">'mirna'</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">X2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># filter out network modules with insufficient number of nodes</span></span>
<span><span class="va">module_length</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">OmicsModule</span>, <span class="va">length</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">network_modules</span> <span class="op">&lt;-</span> <span class="va">OmicsModule</span><span class="op">[</span><span class="va">module_length</span> <span class="op">&gt;</span> <span class="fl">10</span><span class="op">]</span></span>
<span><span class="co"># extract pruned network modules</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">network_modules</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'For network module: '</span>, <span class="va">i</span>, <span class="st">'\n'</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="co"># define subnetwork</span></span>
<span>  <span class="va">abar_sub</span> <span class="op">&lt;-</span> <span class="va">Abar</span><span class="op">[</span><span class="va">network_modules</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>,<span class="va">network_modules</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="va">cor_sub</span> <span class="op">&lt;-</span> <span class="va">bigCor2</span><span class="op">[</span><span class="va">network_modules</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>,<span class="va">network_modules</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="co"># prune network module</span></span>
<span>  <span class="fu"><a href="../reference/networkPruning.html">networkPruning</a></span><span class="op">(</span>Abar <span class="op">=</span> <span class="va">abar_sub</span>,CorrMatrix <span class="op">=</span> <span class="va">cor_sub</span>, </span>
<span>                          type <span class="op">=</span> <span class="va">types</span><span class="op">[</span><span class="va">network_modules</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">]</span>, </span>
<span>                 data <span class="op">=</span> <span class="va">X_big</span><span class="op">[</span>,<span class="va">network_modules</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">]</span>,      </span>
<span>              Pheno <span class="op">=</span> <span class="va">Y</span>, ModuleIdx <span class="op">=</span> <span class="va">i</span>, min_mod_size <span class="op">=</span> <span class="fl">10</span>, </span>
<span>                          max_mod_size <span class="op">=</span> <span class="fl">100</span>, method <span class="op">=</span> <span class="st">'NetSHy'</span>, </span>
<span>                          saving_dir <span class="op">=</span> <span class="va">CVDir</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="results">Results<a class="anchor" aria-label="anchor" href="#results"></a>
</h2>
<p>We present the network result based on the synthetic data. The first
table below contains the individual molecular features correlation with
respect to phenotype, and their associated p-value (from correlation
testing).</p>
<table class="table">
<caption>Individual molecular features correlation table with respect to
phenotype (correlation and p-value).</caption>
<thead><tr class="header">
<th align="left">Molecular Feature</th>
<th align="right">Correlation to Phenotype</th>
<th align="right">P-value</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Gene_1</td>
<td align="right">0.3828730</td>
<td align="right">0.0000000</td>
</tr>
<tr class="even">
<td align="left">Gene_2</td>
<td align="right">0.3411378</td>
<td align="right">0.0000000</td>
</tr>
<tr class="odd">
<td align="left">Gene_5</td>
<td align="right">0.1310036</td>
<td align="right">0.0131111</td>
</tr>
<tr class="even">
<td align="left">Gene_6</td>
<td align="right">0.6284530</td>
<td align="right">0.0000000</td>
</tr>
<tr class="odd">
<td align="left">Gene_7</td>
<td align="right">0.6531262</td>
<td align="right">0.0000000</td>
</tr>
<tr class="even">
<td align="left">Gene_9</td>
<td align="right">0.0774726</td>
<td align="right">0.1434883</td>
</tr>
<tr class="odd">
<td align="left">Gene_10</td>
<td align="right">0.1201825</td>
<td align="right">0.0229506</td>
</tr>
<tr class="even">
<td align="left">Gene_74</td>
<td align="right">-0.1401523</td>
<td align="right">0.0079155</td>
</tr>
<tr class="odd">
<td align="left">Gene_88</td>
<td align="right">0.1234406</td>
<td align="right">0.0194714</td>
</tr>
<tr class="even">
<td align="left">Gene_90</td>
<td align="right">-0.0752956</td>
<td align="right">0.1551147</td>
</tr>
<tr class="odd">
<td align="left">Gene_123</td>
<td align="right">-0.0703372</td>
<td align="right">0.1842334</td>
</tr>
<tr class="even">
<td align="left">Gene_170</td>
<td align="right">0.1007365</td>
<td align="right">0.0568835</td>
</tr>
<tr class="odd">
<td align="left">Gene_216</td>
<td align="right">-0.0946349</td>
<td align="right">0.0737229</td>
</tr>
<tr class="even">
<td align="left">Gene_219</td>
<td align="right">0.1158386</td>
<td align="right">0.0284166</td>
</tr>
<tr class="odd">
<td align="left">Gene_367</td>
<td align="right">0.1499230</td>
<td align="right">0.0044714</td>
</tr>
<tr class="even">
<td align="left">Gene_391</td>
<td align="right">-0.0570075</td>
<td align="right">0.2820460</td>
</tr>
<tr class="odd">
<td align="left">Gene_485</td>
<td align="right">0.1176494</td>
<td align="right">0.0260153</td>
</tr>
<tr class="even">
<td align="left">Mir_2</td>
<td align="right">-0.3474835</td>
<td align="right">0.0000000</td>
</tr>
</tbody>
</table>
<div class="figure">
<img src="SmCCNet_Vignette_MultiOmics_files/figure-html/loadings1-1.png" alt="PC1 loading for each subnetwork feature." width="100%"><p class="caption">
PC1 loading for each subnetwork feature.
</p>
</div>
<div class="figure">
<img src="SmCCNet_Vignette_MultiOmics_files/figure-html/corheatmap-1.png" alt="Correlation heatmap for subnetwork features." width="100%"><p class="caption">
Correlation heatmap for subnetwork features.
</p>
</div>
<div class="figure">
<img src="SmCCNet_Vignette_MultiOmics_files/figure-html/adjheatmap-1.png" alt="Adjacency matrix heatmap for subnetwork features." width="100%"><p class="caption">
Adjacency matrix heatmap for subnetwork features.
</p>
</div>
<p>Figure <span class="math inline">\(\ref{fig:loadings1}\)</span> is
the visualization of the PC loadings that contribution of each molecular
features to the first NetSHy PC. In addition, there are two network
heatmaps based on (1) correlation matrix (Figure <span class="math inline">\(\ref{fig:corheatmap}\)</span>), and (2) adjacency
matrix (Figure <span class="math inline">\(\ref{fig:adjheatmap}\)</span>). Based on the
summarization table, genes 1,2,6,7, and miRNA 2 have relatively high
correlation with respect to phenotype. The PC loadings also confirm that
genes 6,7, and miRNA 2 and generally have higher PC contribution. From
the correlation heatmap, we do not observe associations between
molecular features, but for the adjacency matrix heatmap, we observe the
higher connections between genes 6,7 and miRNA 2. We also recommend
experimenting more PCs (up to 3) with Figure <span class="math inline">\(\ref{fig:loadings1}\)</span> to demonstrate the
contribution of molecular features to each PC.</p>
<div class="section level3">
<h3 id="step-vi-visualize-network-module">Step VI: Visualize network module<a class="anchor" aria-label="anchor" href="#step-vi-visualize-network-module"></a>
</h3>
<p>The initial approach to network visualization is facilitated through
our SmCCNet shinyApp, accessible at <a href="https://smccnet.shinyapps.io/smccnetnetwork/" class="external-link">https://smccnet.shinyapps.io/smccnetnetwork/</a>.
Upon obtaining a subnetwork file named ‘size_a_net_b.Rdata’, users can
upload it to the shinyApp. The platform provides various adjustable
visualization parameters, enabling users to tailor the network
visualization to their preferences.</p>
<p>An alternative way to visualize the final network module, we need to
download the Cytoscape software (Shannon et al., 2003), and use the
package RCy3 to visualize the subnetwork generated from the network
pruning step. In general, since the network obtained through the
PageRank pruning algorithm is densely connected, and some of the edges
may be false positive (meaning that two nodes are not associated, but
with higher edge values in the adjacency matrix). Therefore, we use the
correlation matrix to filter out those weak network edges.</p>
<p>In the network visualization (Figure <span class="math inline">\(\ref{fig:netPlot}\)</span>), different colored
edges denote different directions of the association between two nodes,
where red or blue denotes a positive or negative association
respectively. The width of the edge represents the connection strength
between two nodes.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">RCy3</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r.igraph.org/" class="external-link">igraph</a></span><span class="op">)</span></span>
<span><span class="co"># load subnetwork data (example, user need to provide the directory)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="st">'ResultDirectory/size_a_net_b.Rdata'</span><span class="op">)</span></span>
<span><span class="va">M</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">M</span><span class="op">)</span></span>
<span><span class="va">correlation_filter</span> <span class="op">&lt;-</span> <span class="fl">0.05</span></span>
<span><span class="co"># correlation matrix filtering for the subnetwork edge-cut</span></span>
<span><span class="va">filter_index</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">correlation_sub</span><span class="op">)</span> <span class="op">&lt;</span> <span class="va">correlation_filter</span><span class="op">)</span></span>
<span><span class="va">M_ind</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">correlation_sub</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="fl">1</span>, <span class="op">-</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">M_adj</span> <span class="op">&lt;-</span> <span class="va">M</span> <span class="op">*</span> <span class="va">M_ind</span></span>
<span><span class="va">M_adj</span><span class="op">[</span><span class="va">filter_index</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">M_adj</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span></span>
<span><span class="co"># network visualization through cytoscape</span></span>
<span><span class="va">graph</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/graph_from_adjacency_matrix.html" class="external-link">graph_from_adjacency_matrix</a></span><span class="op">(</span><span class="va">M_adj</span>, mode <span class="op">=</span> <span class="st">'undirected'</span>, </span>
<span>         weighted <span class="op">=</span> <span class="cn">TRUE</span>, diag <span class="op">=</span> <span class="cn">TRUE</span>, add.colnames <span class="op">=</span> <span class="cn">NULL</span>, add.rownames <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># define network node type and connectivity and use them in cytoscape</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">type</span> <span class="op">&lt;-</span> <span class="va">sub_type</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">type</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">connectivity</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">M</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">connectivity</span></span>
<span><span class="co"># export subnetwork to Cytoscape </span></span>
<span><span class="fu">createNetworkFromIgraph</span><span class="op">(</span><span class="va">graph</span>,<span class="st">"multi_omics_network"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="../articles/figures/MultiOmicsNet.png" alt="Pruned module 1. The strength of the node connections is indicated by the thickness of edges. Red edges and blue edges are for negative and positive connections respectively. Red node represents genes, and blue node represent miRNAs." width="100%"><p class="caption">
Pruned module 1. The strength of the node connections is indicated by
the thickness of edges. Red edges and blue edges are for negative and
positive connections respectively. Red node represents genes, and blue
node represent miRNAs.
</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="methods-for-optimal-scaling-factors-selection">Methods for Optimal Scaling Factors Selection<a class="anchor" aria-label="anchor" href="#methods-for-optimal-scaling-factors-selection"></a>
</h2>
<p>As shown above, the scaling factors must be supplied to prioritize
correlation structure of interest. However, it is not always
straightforward to determine the scaling factors. In the section below,
we provide 3 different methods to select the scaling factors of
interest.</p>
<div class="section level3">
<h3 id="method-1-use-prompt-to-define-scaling-factors">Method 1: Use Prompt to Define Scaling Factors<a class="anchor" aria-label="anchor" href="#method-1-use-prompt-to-define-scaling-factors"></a>
</h3>
<p>A common choice for the scaling factors is to prioritize the
omics-phenotype correlation. For instance, in multi-omics quantitative
phenotype analysis with 2 omics data, let <span class="math inline">\(a_{1,2}\)</span> be the scaling factor for
omics-omics correlation, <span class="math inline">\(b_{1}\)</span> and
<span class="math inline">\(b_{2}\)</span> be the scaling factors for
omics-phenotype correlation, then a common choice of the scaling factor
<span class="math inline">\((a_{1,2},b_1,b_2)\)</span> is <span class="math inline">\((1,10,10)\)</span>. This can be done through
manual entry with interactive function ‘scalingFactorInput()’, and the
example is given in section 2.2.2.</p>
</div>
<div class="section level3">
<h3 id="method-2-use-automated-smccnet-to-select-scaling-factors-based-on-pairwise-correlation-">Method 2: Use Automated SmCCNet to Select Scaling Factors Based on
Pairwise Correlation.<a class="anchor" aria-label="anchor" href="#method-2-use-automated-smccnet-to-select-scaling-factors-based-on-pairwise-correlation-"></a>
</h3>
<p>Another method is to check the pairwise correlation between each pair
of omics data using sparse canonical correlation analysis (SCCA). We ran
SCCA with the most stringent penalty pair (<span class="math inline">\(0.1,0.1\)</span>), and calculate the canonical
correlation. The canonical correlation calculated will be treated as the
between-omics scaling factor, while scaling factor of 1 will be used for
omics-phenotype relationship. In addition, we introduced another
parameter called shrinkage factor to prioritize either omics-omics
relationship or omics-phenotype relationship. For instance, in a
multi-omics analysis with two omics data, if omics-omics correlation is
0.8, and the shrinkage parameter is 2, then the final scaling factors
are set to <span class="math inline">\((a,b_1,b_2) = c(0.4, 1,
1)\)</span>. This method is currently only implemented in
<strong>fastAutoSmCCNet()</strong>. If user want to use this method,
please refer to the automated SmCCNet vignette.</p>
</div>
<div class="section level3">
<h3 id="method-3-use-cross-validation-to-select-scaling-factors">Method 3: Use Cross-Validation to Select Scaling Factors<a class="anchor" aria-label="anchor" href="#method-3-use-cross-validation-to-select-scaling-factors"></a>
</h3>
<p>Let’s assume that there are two omics data mRNA and miRNA with a
quantitative phenotype. We want to systematically evaluate the
performance of each pair of scaling factors. The best way is to create
candidate scaling factors set by setting all omics-omics scaling factor
to 1, and varying the omics-phenotype scaling factors. To make
performance comparable, we scale each set of scaling factors so that
they sum up to 1. For instance, with the two omics example, the scaling
factor <span class="math inline">\((a_{1,2}, b_1, b_2)\)</span> should
satisfy <span class="math inline">\(a_{1,2} + b_1 + b_2 = 1\)</span>. We
implement a nested grid search approach to select the best scaling
factors along with the corresponding optimal penalty parameters. For
each set of scaling factors, the sparse penalty parameters (l1, l2) were
chosen through a K-fold cross validation to find the penalty pair that
minimized the scaled prediction error, and the best scaled prediction
error is recorded for each set of scaling factors and is used to find
the optimal scaling factors. All penalty pairs from the set were also
tested in a grid search to find the optimal pair (l1, l2). The penalty
parameters determine how many miRNA (l1) and mRNA (l2) are in the final
results. Scaling factors in the grid search were evaluated to identify
which value yielded the best network results with a predefined
criterion. We discuss the evaluation criterion in section 4.3.2.</p>
<div class="section level4">
<h4 id="cross-validation-with-each-set-of-scaling-factors">Cross validation with each set of scaling factors<a class="anchor" aria-label="anchor" href="#cross-validation-with-each-set-of-scaling-factors"></a>
</h4>
<p>Below is the penalty parameter grid search setup for the SmCCNet
algorithm. Note that it is the same step as the section 2.2:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># number of folds in K-fold CV.</span></span>
<span><span class="va">K</span> <span class="op">&lt;-</span> <span class="fl">3</span> </span>
<span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">X1</span><span class="op">)</span></span>
<span><span class="co"># create a list of omics data</span></span>
<span><span class="va">X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">X1</span>, <span class="va">X2</span><span class="op">)</span></span>
<span><span class="co"># number of omics dataset</span></span>
<span><span class="va">num_omics</span> <span class="op">&lt;-</span> <span class="fl">2</span></span>
<span><span class="co"># tuning parameter candidate length for each omics data</span></span>
<span><span class="va">tuneLength</span> <span class="op">&lt;-</span> <span class="fl">5</span></span>
<span><span class="co"># tuning parameter candadate range for each omics data</span></span>
<span><span class="va">minTune</span> <span class="op">&lt;-</span> <span class="fl">0.1</span></span>
<span><span class="va">maxTune</span> <span class="op">&lt;-</span> <span class="fl">0.5</span></span>
<span><span class="co"># create empty matrix to store all possible penalty parameters</span></span>
<span><span class="va">penSelect</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">0</span>, nrow <span class="op">=</span> <span class="va">tuneLength</span>, ncol <span class="op">=</span> <span class="va">num_omics</span><span class="op">)</span></span>
<span><span class="co"># create sparsity penalty options.</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">Idx</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">penSelect</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">{</span></span>
<span>  <span class="va">penSelect</span><span class="op">[</span>,<span class="va">Idx</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span>from <span class="op">=</span> <span class="va">minTune</span>, </span>
<span>                         to <span class="op">=</span> <span class="va">maxTune</span>, </span>
<span>                         length.out <span class="op">=</span> <span class="va">tuneLength</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co"># expand grid</span></span>
<span><span class="co"># convert matrix to list of columns</span></span>
<span><span class="va">list_cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">as.list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">penSelect</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># generate all possible combinations</span></span>
<span><span class="va">PenExpand</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">expand.grid</span>, <span class="va">list_cols</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># set a CV directory.</span></span>
<span><span class="va">CVDir</span> <span class="op">&lt;-</span> <span class="st">"Example3foldCVTune/"</span> </span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">CVDir</span><span class="op">)</span></span></code></pre></div>
<p>Same as above, we split the original data into K different folds for
evaluation:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">12345</span><span class="op">)</span> <span class="co"># set random seed.</span></span>
<span></span>
<span><span class="co"># split data into folds</span></span>
<span><span class="va">X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">X</span>, <span class="va">scale</span><span class="op">)</span></span>
<span><span class="va">foldIdx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">X</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">X</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, <span class="va">K</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">folddata</span> <span class="op">&lt;-</span> <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">foldIdx</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">Y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span></span>
<span>    <span class="va">X_train</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="va">X_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="va">Y_train</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="va">Y_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">{</span></span>
<span>      <span class="va">X_train</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">X</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="op">-</span><span class="va">foldIdx</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span>,<span class="op">]</span></span>
<span>      <span class="va">X_test</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">X</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">foldIdx</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span>,<span class="op">]</span></span>
<span>    <span class="op">}</span></span>
<span>    <span class="va">Y_train</span> <span class="op">&lt;-</span> <span class="va">Y</span><span class="op">[</span><span class="op">-</span><span class="va">foldIdx</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span>,<span class="op">]</span></span>
<span>    <span class="va">Y_test</span> <span class="op">&lt;-</span> <span class="va">Y</span><span class="op">[</span><span class="va">foldIdx</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span>,<span class="op">]</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>X_train <span class="op">=</span> <span class="va">X_train</span>, X_test <span class="op">=</span> <span class="va">X_test</span>,Y_train <span class="op">=</span> <span class="va">Y_train</span>, </span>
<span>                Y_test <span class="op">=</span> <span class="va">Y_test</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="co"># name each fold of data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">folddata</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'Fold_'</span>, <span class="fl">1</span><span class="op">:</span><span class="va">K</span><span class="op">)</span></span>
<span><span class="co"># saving all preliminary data into local directory for reproducibility purpose</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span><span class="va">folddata</span>, <span class="va">PenExpand</span>,</span>
<span>       file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">CVDir</span>, <span class="st">"CVData.RData"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Below is the function to set up the searching grid for scaling
factors. There are 3 different arguments in this function:</p>
<ul>
<li>
<span class="math inline">\(DataType\)</span>: The vector with each
omics type annotated (in same order as <span class="math inline">\(X\)</span>), followed by ‘phenotype’.</li>
<li>
<span class="math inline">\(tuneLength\)</span>: The length of
candidate scaling factors for each pairwise combination.</li>
<li>
<span class="math inline">\(tuneRangePheno\)</span>: A vector of
length 2 with the first argument being the minimally possible scaling
factor value, and the second argument being the maximally possible
scaling factor value.</li>
</ul>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create a function that set up the scaling factor candidate grids</span></span>
<span><span class="va">gridCCcoef</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">DataType</span>, <span class="va">tuneLength</span> <span class="op">=</span> <span class="fl">5</span>, <span class="va">tuneRangePheno</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">{</span></span>
<span>  <span class="co"># store the length of the data</span></span>
<span>  <span class="va">datalength</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">DataType</span><span class="op">)</span></span>
<span>  <span class="va">phenotunelength</span> <span class="op">&lt;-</span> <span class="va">datalength</span> <span class="op">-</span> <span class="fl">1</span></span>
<span>  <span class="co"># create empty matrix for storing the candidate scaling factors</span></span>
<span>  <span class="va">candCoef</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">1</span>, nrow <span class="op">=</span> <span class="va">tuneLength</span><span class="op">^</span><span class="va">phenotunelength</span>, </span>
<span>                       ncol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/combn.html" class="external-link">combn</a></span><span class="op">(</span><span class="va">datalength</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="co"># create storage empty grid</span></span>
<span>  <span class="va">phenoGrids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">0</span>, nrow <span class="op">=</span> <span class="va">tuneLength</span>, </span>
<span>                         ncol <span class="op">=</span> <span class="va">phenotunelength</span><span class="op">)</span></span>
<span>  <span class="co"># create grids</span></span>
<span>  <span class="kw">for</span> <span class="op">(</span><span class="va">phenoIdx</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">phenotunelength</span><span class="op">)</span></span>
<span>  <span class="op">{</span></span>
<span>    <span class="va">phenoGrids</span><span class="op">[</span>,<span class="va">phenoIdx</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span>from <span class="op">=</span> <span class="va">tuneRangePheno</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, </span>
<span>                                 to <span class="op">=</span> <span class="va">tuneRangePheno</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, </span>
<span>                                 length.out <span class="op">=</span> <span class="va">tuneLength</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="co"># expand grid</span></span>
<span>  <span class="co"># convert matrix to list of columns</span></span>
<span>  <span class="va">list_cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">as.list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">phenoGrids</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="co"># generate combinations</span></span>
<span>  <span class="va">phenoGridsExpand</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">expand.grid</span>, <span class="va">list_cols</span><span class="op">)</span></span>
<span>  <span class="va">candCoef</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/combn.html" class="external-link">combn</a></span><span class="op">(</span><span class="va">datalength</span>,<span class="fl">2</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span>,<span class="op">]</span> <span class="op">==</span> <span class="va">datalength</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span></span>
<span>    <span class="va">phenoGridsExpand</span><span class="op">)</span></span>
<span>  <span class="co"># provide column names</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">candCoef</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/combn.html" class="external-link">combn</a></span><span class="op">(</span><span class="va">datalength</span>,<span class="fl">2</span><span class="op">)</span>,<span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">DataType</span><span class="op">[</span><span class="va">x</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,<span class="st">'_'</span>, <span class="va">DataType</span><span class="op">[</span><span class="va">x</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>    <span class="op">}</span><span class="op">)</span></span>
<span>  <span class="co"># scale scaling factors so that each set of scaling factors sum up to 1</span></span>
<span>  <span class="va">candCoef</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">candCoef</span>, <span class="fl">1</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span><span class="va">x</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">candCoef</span><span class="op">)</span></span>
<span>    </span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># scaling factor grids</span></span>
<span><span class="va">CCcoefMatrix</span> <span class="op">&lt;-</span> <span class="fu">gridCCcoef</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'mRNA'</span>,<span class="st">'miRNA'</span>, <span class="st">'pheno'</span><span class="op">)</span>, </span>
<span>                           tuneLength <span class="op">=</span> <span class="fl">3</span>, tuneRangePheno <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>,<span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># create data matrix to store the cross-validation result</span></span>
<span><span class="va">CVEval</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">0</span>, nrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">CCcoefMatrix</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="va">num_omics</span> <span class="op">+</span> </span>
<span>                   <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/combn.html" class="external-link">combn</a></span><span class="op">(</span><span class="va">num_omics</span> <span class="op">+</span> <span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">CVEval</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/combn.html" class="external-link">combn</a></span><span class="op">(</span><span class="va">num_omics</span> <span class="op">+</span> <span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">CCcoefMatrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">CVEval</span><span class="op">)</span><span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'CCcoef:'</span>,  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">CCcoefMatrix</span><span class="op">)</span><span class="op">)</span>, </span>
<span>                      <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'l'</span>,<span class="fl">1</span><span class="op">:</span><span class="va">num_omics</span><span class="op">)</span>, <span class="st">"TrainingCC"</span>, <span class="st">"TestCC"</span>, </span>
<span>                                <span class="st">"CCPredError"</span><span class="op">)</span></span></code></pre></div>
<p>In the example below, with two omics and a quantitative phenotype, we
set the <span class="math inline">\(tuneLength\)</span> to 3, with the
range of scaling factor to be between 5 and 10. Since there are a large
number of candidate scaling factor pairs, to speed up the
cross-validation step, the parallel computation can be implemented.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">for</span> <span class="op">(</span><span class="va">CCIdx</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">CCcoefMatrix</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">{</span></span>
<span>  <span class="va">coef</span> <span class="op">&lt;-</span> <span class="va">CCcoefMatrix</span><span class="op">[</span><span class="va">CCIdx</span>,<span class="op">]</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'Now running SmCCA for the scaling factor candidate '</span>, </span>
<span>             <span class="va">CCIdx</span>, <span class="st">'\n'</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>  <span class="fu">future</span><span class="fu">::</span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="fu">future</span><span class="fu">::</span><span class="va"><a href="https://future.futureverse.org/reference/multisession.html" class="external-link">multisession</a></span>, workers <span class="op">=</span> <span class="va">K</span><span class="op">)</span></span>
<span>  <span class="va">CVResult</span> <span class="op">&lt;-</span> <span class="fu">furrr</span><span class="fu">::</span><span class="fu"><a href="https://furrr.futureverse.org/reference/future_map.html" class="external-link">future_map</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">K</span>, <span class="kw">function</span><span class="op">(</span><span class="va">CVidx</span><span class="op">)</span> <span class="op">{</span>  </span>
<span>    <span class="co"># set scaling factor</span></span>
<span>    <span class="va">CCcoef</span> <span class="op">&lt;-</span> <span class="va">coef</span></span>
<span>    <span class="co"># create empty vector for storing cross-validation result</span></span>
<span>    <span class="va">RhoTrain</span> <span class="op">&lt;-</span> <span class="va">RhoTest</span> <span class="op">&lt;-</span> <span class="va">DeltaCor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">PenExpand</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="kw">for</span><span class="op">(</span><span class="va">idx</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">PenExpand</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">{</span></span>
<span>      <span class="co"># consider one pair of sparsity penalties at a time.</span></span>
<span>      <span class="va">l</span> <span class="op">&lt;-</span> <span class="va">PenExpand</span><span class="op">[</span><span class="va">idx</span>, <span class="op">]</span></span>
<span>      <span class="co"># run SmCCA on the subsamples (Figure 1, Step II)</span></span>
<span>      <span class="va">Ws</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getCanWeightsMulti.html">getCanWeightsMulti</a></span><span class="op">(</span><span class="va">folddata</span><span class="op">[[</span><span class="va">CVidx</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"X_train"</span><span class="op">]</span><span class="op">]</span>, </span>
<span>                           Trait <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">folddata</span><span class="op">[[</span><span class="va">CVidx</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"Y_train"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, </span>
<span>                           Lambda <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">l</span><span class="op">)</span>, NoTrait <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                           CCcoef <span class="op">=</span> <span class="va">CCcoef</span><span class="op">)</span></span>
<span>      <span class="co"># evaluate the canonical correlation for training and testing data</span></span>
<span>      <span class="va">rho.train</span> <span class="op">&lt;-</span>  <span class="fu"><a href="../reference/getCanCorMulti.html">getCanCorMulti</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">folddata</span><span class="op">[[</span><span class="va">CVidx</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"X_train"</span><span class="op">]</span><span class="op">]</span>, </span>
<span>                                Y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">folddata</span><span class="op">[[</span><span class="va">CVidx</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"Y_train"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, </span>
<span>                                CCWeight <span class="op">=</span> <span class="va">Ws</span>, </span>
<span>                                CCcoef <span class="op">=</span> <span class="va">CCcoef</span><span class="op">)</span></span>
<span>      <span class="va">rho.test</span> <span class="op">&lt;-</span>  <span class="fu"><a href="../reference/getCanCorMulti.html">getCanCorMulti</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">folddata</span><span class="op">[[</span><span class="va">CVidx</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"X_test"</span><span class="op">]</span><span class="op">]</span>, </span>
<span>                                Y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">folddata</span><span class="op">[[</span><span class="va">CVidx</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"Y_test"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, </span>
<span>                                CCWeight <span class="op">=</span> <span class="va">Ws</span>, </span>
<span>                                CCcoef <span class="op">=</span> <span class="va">CCcoef</span><span class="op">)</span></span>
<span></span>
<span>      <span class="co"># store cv result  </span></span>
<span>      <span class="va">RhoTrain</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">rho.train</span>, digits <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span>      <span class="va">RhoTest</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">rho.test</span>, digits <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span>      <span class="va">DeltaCor</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">rho.train</span> <span class="op">-</span> <span class="va">rho.test</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="op">}</span></span>
<span>    </span>
<span>    <span class="co"># record prediction errors for given CV fold and all sparsity penalty </span></span>
<span>    <span class="co"># options.</span></span>
<span>    <span class="va">DeltaCor.all</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">RhoTrain</span>, <span class="va">RhoTest</span>, <span class="va">DeltaCor</span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">DeltaCor.all</span><span class="op">)</span></span>
<span>  <span class="op">}</span>,.progress <span class="op">=</span> <span class="cn">TRUE</span>,.options <span class="op">=</span> <span class="fu">furrr</span><span class="fu">::</span><span class="fu"><a href="https://furrr.futureverse.org/reference/furrr_options.html" class="external-link">furrr_options</a></span><span class="op">(</span>seed <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">'\n'</span><span class="op">)</span></span>
<span>  <span class="co"># aggregate CV result and select the best penalty term</span></span>
<span>  <span class="va">AggregatedCVResult</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/funprog.html" class="external-link">Reduce</a></span><span class="op">(</span><span class="st">"+"</span>, <span class="va">CVResult</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">CVResult</span><span class="op">)</span></span>
<span>  <span class="va">EvalMetric</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">AggregatedCVResult</span>, <span class="fl">1</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span><span class="va">x</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">x</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></span>
<span>  <span class="co"># determine the best CV result</span></span>
<span>  <span class="va">optIdx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.min.html" class="external-link">which.min</a></span><span class="op">(</span><span class="va">EvalMetric</span><span class="op">)</span></span>
<span>  <span class="co"># fill in the optimal penalty pair for current scaling scaling </span></span>
<span>  <span class="co"># factor selection as well as the evaluation result</span></span>
<span>  <span class="va">CVEval</span><span class="op">[</span><span class="va">CCIdx</span>,<span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/combn.html" class="external-link">combn</a></span><span class="op">(</span><span class="va">num_omics</span> <span class="op">+</span> <span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>               <span class="fl">1</span><span class="op">)</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">CVEval</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">PenExpand</span><span class="op">[</span><span class="va">optIdx</span>,<span class="op">]</span><span class="op">)</span>,</span>
<span>                      <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">AggregatedCVResult</span><span class="op">[</span><span class="va">optIdx</span>,<span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># write out the cross-validation result</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.csv</a></span><span class="op">(</span><span class="va">CVEval</span>, </span>
<span>            file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">CVDir</span>, <span class="st">"/PredictionError.csv"</span><span class="op">)</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="select-optimal-scaling-factors-with-associated-penalty-parameters">Select optimal scaling factors with associated penalty
parameters<a class="anchor" aria-label="anchor" href="#select-optimal-scaling-factors-with-associated-penalty-parameters"></a>
</h4>
<p>Scaling constants in the grid search were evaluated to identify which
values yielded the best network results. How to define a evaluation
criterion depends on the interest of study. In the two omics example,
the criterion is set to be:</p>
<p><span class="math display">\[\begin{eqnarray}
\begin{split}
\label{eq:scaling}
&amp; (a_{1,2}, b_1, b_2)\!  =\! \arg\max_{\tilde a_{1,2}, \tilde b_1,
\tilde c_2} \frac{|trainCC - testCC|}{|testCC|}  \end{split}
\end{eqnarray}\]</span></p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># read in the overall cv result</span></span>
<span><span class="va">evalResult</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">CVDir</span>, <span class="st">"/PredictionError.csv"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># find the optn</span></span>
<span><span class="va">evalOptIdx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.min.html" class="external-link">which.min</a></span><span class="op">(</span><span class="va">evalResult</span><span class="op">$</span><span class="va">CCPredError</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">evalResult</span><span class="op">$</span><span class="va">TestCC</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># print the optimal result </span></span>
<span><span class="va">evalResult</span><span class="op">[</span><span class="va">evalOptIdx</span>,<span class="op">]</span></span></code></pre></div>
<p>With the selected parameters of scaling factors (CCcoef), l1, and l2,
we run SmCCNet on the entire dataset as Session 2.3 and identify the
networks related to phenotype of interest as session 2.4.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># feature sampling proportions, 0.9 for miRNA since it has less features.</span></span>
<span><span class="va">s</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.7</span>,<span class="fl">0.9</span><span class="op">)</span> </span>
<span><span class="co"># number of subsamples.</span></span>
<span><span class="va">SubsamplingNum</span> <span class="op">&lt;-</span> <span class="fl">50</span></span>
<span><span class="co"># run SmCCA on the subsamples (Figure 1, Step II)</span></span>
<span><span class="va">Ws</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getRobustWeightsMulti.html">getRobustWeightsMulti</a></span><span class="op">(</span><span class="va">X</span>, </span>
<span>                        Trait <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span>, </span>
<span>                        NoTrait <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                        CCcoef <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">evalResult</span><span class="op">[</span><span class="va">evalOptIdx</span>, </span>
<span>                                 <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/combn.html" class="external-link">combn</a></span><span class="op">(</span><span class="va">num_omics</span> <span class="op">+</span> <span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                        Lambda <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">evalResult</span><span class="op">[</span><span class="va">evalOptIdx</span>,</span>
<span>                                <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/combn.html" class="external-link">combn</a></span><span class="op">(</span><span class="va">num_omics</span> <span class="op">+</span> <span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span>                                 <span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">:</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/combn.html" class="external-link">combn</a></span><span class="op">(</span><span class="va">num_omics</span> <span class="op">+</span> </span>
<span>                                 <span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">+</span><span class="va">num_omics</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>, s <span class="op">=</span> <span class="va">s</span>, </span>
<span>                        SubsamplingNum <span class="op">=</span> <span class="va">SubsamplingNum</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="multi-omics-smccnet-with-binary-phenotype">Multi-Omics SmCCNet with Binary Phenotype<a class="anchor" aria-label="anchor" href="#multi-omics-smccnet-with-binary-phenotype"></a>
</h2>
<div class="figure">
<img src="figures/SmCCNet-Binary.jpg" alt="SmCCNet workflow overview for Binary Phenotype. X1, X2, and X3 are three omics data types for the same set of n subjects. Y indicates a Binary phenotype measure for those n subjects. Note that the flowchart demonstrate workflow for three omics data, it is also compatible with more than three omics data or two omics data." width="100%"><p class="caption">
SmCCNet workflow overview for Binary Phenotype. X1, X2, and X3 are three
omics data types for the same set of n subjects. Y indicates a Binary
phenotype measure for those n subjects. Note that the flowchart
demonstrate workflow for three omics data, it is also compatible with
more than three omics data or two omics data.
</p>
</div>
<p>Consider <span class="math inline">\(X_1, X_2, ..., X_T\)</span> as
<span class="math inline">\(T\)</span> omics datasets, and <span class="math inline">\(Y\)</span> as the phenotype data. The general
workflow for multi-omics SmCCNet with binary phenotype is as follows
(Figure <span class="math inline">\(\ref{fig:flowchart-binary}\)</span>):</p>
<ol style="list-style-type: decimal">
<li><p><strong>Run Weighted/Unweighted Sparse Multiple Canonical
Correlation Analysis (SmCCA)</strong>: This is done on <span class="math inline">\(X_1, X_2, ..., X_T\)</span> (excluding phenotype
data). The output is canonical weight vectors (with nonzero entries,
zero entries are filtered) <span class="math inline">\(\tilde W_{t} \in
\mathbb{R}^{p_t^{(sub)} \times 1}, t = 1,2,...,T\)</span>, which
represent the omics-omics connections. In this step, we filter out
features that have no connections with other features, which helps
reduce dimensionality. Note that we tend to set relaxed penalty terms
for this step to include as many omics features as possible to increase
the performance of the classifier in the next step.</p></li>
<li><p><strong>Subset Omics Data</strong>: Each dataset <span class="math inline">\(X_1, X_2, ..., X_T\)</span> is subsetted to
include only omics features selected in Step 1, call subsetted data
<span class="math inline">\(X_t(sub)\in \mathbb{R}^{n \times
p_{t}^{(sub)}}\)</span>.</p></li>
<li><p><strong>Concatenate and Run Sparse Partial Least Squared
Discriminant Analysis (SPLSDA) (Chung and Keles, 2010)</strong>: The
subsetted datasets <span class="math inline">\(X_1(sub), X_2(sub),...,
X_T(sub)\)</span> are concatenated into <span class="math inline">\(X(sub) = [X_1(sub), X_2(sub),...,X_T(sub)] \in
\mathbb{R}^{n\times p}, p = \sum_{i = 1}^{T} p_i\)</span>. The Sparse
PLSDA algorithm is then run to extract <span class="math inline">\(R\)</span> latent factors and projection matrix,
by default, <span class="math inline">\(R\)</span> is set to 3.
Projection matrix is defined as <span class="math inline">\(P \in
\mathbb{R}^{p^{(sub)} \times R}\)</span>. Latent factors are defined as
<span class="math inline">\(r = [r_1,r_2,...,r_R] = X(sub) \cdot P \in
\mathbb{R}^{n \times R}\)</span>.</p></li>
<li><p><strong>Aggregate Latent Factors</strong>: The <span class="math inline">\(R\)</span> latent factors are aggregated into one
using logistic regression, defined by logit<span class="math inline">\((Y) = \alpha_1 r_1 + \alpha_2 r_2 + ... +\alpha_R
r_R\)</span>. Estimate <span class="math inline">\(\alpha\)</span>’s and
report performance (e.g., accuracy/AUC score). Feature weights are given
by aggregation of the projection matrix from Sparse PLSDA <span class="math inline">\(W_{t}^{*} = P_{t} \cdot \alpha \in
\mathbb{R}^{p_t^{(sub)} \times 1}, t = 1,2,...,T, \alpha = [\alpha_1,
\alpha_2,....,\alpha_r] \in \mathbb{R}^{R \times 1}\)</span>, where
<span class="math inline">\(P_{t}\)</span> is the subset of projection
matrix <span class="math inline">\(P\)</span> such that it only includes
features from the <span class="math inline">\(t\)</span>th omics
data.</p></li>
<li><p><strong>Normalize and Calculate Final Canonical Weight</strong>:
The feature weights <span class="math inline">\(W_{1}^{*},W_{2}^{*},...,W_{T}^{*}\)</span> based
on sparse PLSDA are normalized to have an L2 norm of 1. Let <span class="math inline">\(\gamma_1\)</span> and <span class="math inline">\(\gamma_2\)</span> be two scalars representing the
strength of omics-omics and omics-phenotype connections,
respectively,then the final canonical weight is obtained by weighted
combining the canonical weight from step 1 and the feature weight from
classifier <span class="math inline">\(W_{t} =
\frac{\gamma_1}{\gamma_1+\gamma_2} \tilde W_{t} +
\frac{\gamma_2}{\gamma_1+\gamma_2} W_{t}^{*}, t =
1,2,...,T\)</span>.</p></li>
<li><p><strong>Construct Multi-Omics Network and Perform Network
Pruning</strong>: A multi-omics network is constructed and pruned (same
as multi-omics SmCCNet with quantitative phenotype).</p></li>
</ol>
<p>Note that Steps 1-3 are repeated multiple times with different
penalty pairs to select the best pair with cross-validation. The
evaluation metric is maximizing the classification evaluation metric
such as testing prediction accuracy or testing AUC score. Currently
there are 5 different evaluation metrics to choose from: accuracy, AUC
score, precision, recall, and F1 score, with accuracy as the default
selection. For instance, if the phenotype is subject’s disease status
with possible values of 1 (disease-positive) and 0 (disease-negative),
then below is the guidance on how to choose the best evaluation metric
in this context:</p>
<div class="section level3">
<h3 id="synthetic-dataset-1">Synthetic dataset<a class="anchor" aria-label="anchor" href="#synthetic-dataset-1"></a>
</h3>
<p>As an example, we consider a synthetic data set with 500 genes <span class="math inline">\((X_1)\)</span> and 100 miRNAs <span class="math inline">\((X_2)\)</span> expression levels measured for 358
subjects, along with a binary phenotype created by binarizing
quantitative phenotype <span class="math inline">\((Y)\)</span> based on
median.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">ExampleData</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">X1</span><span class="op">[</span> , <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          Gene_1   Gene_2   Gene_3   Gene_4   Gene_5   Gene_6</span></span>
<span><span class="co">## Samp_1 22.48570 40.35372 31.02575 20.84721 26.69729 30.20545</span></span>
<span><span class="co">## Samp_2 37.05885 34.05223 33.48702 23.53146 26.75463 31.73594</span></span>
<span><span class="co">## Samp_3 20.53077 31.66962 35.18957 20.95254 25.01883 32.15723</span></span>
<span><span class="co">## Samp_4 33.18689 38.48088 18.89710 31.82330 34.04938 38.79989</span></span>
<span><span class="co">## Samp_5 28.96198 41.06049 28.49496 18.37449 30.81524 24.00454</span></span>
<span><span class="co">## Samp_6 18.05983 29.55471 32.54002 29.68452 26.19996 26.76684</span></span></code></pre>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">X2</span><span class="op">[</span> , <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##           Mir_1    Mir_2    Mir_3    Mir_4    Mir_5     Mir_6</span></span>
<span><span class="co">## Samp_1 15.22391 17.54583 15.78472 14.89198 10.34821  9.689755</span></span>
<span><span class="co">## Samp_2 16.30697 16.67283 13.36153 14.48855 12.66090 11.333613</span></span>
<span><span class="co">## Samp_3 16.54512 16.73501 14.61747 17.84527 13.82279 11.329333</span></span>
<span><span class="co">## Samp_4 13.98690 16.20743 16.29308 17.72529 12.30056  9.844108</span></span>
<span><span class="co">## Samp_5 16.33833 17.39387 16.39792 15.85373 13.38767 10.599414</span></span>
<span><span class="co">## Samp_6 14.54110 16.51999 14.73958 15.87504 13.21359 10.922393</span></span></code></pre>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># binarize phenotype variable</span></span>
<span><span class="va">Y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">Y</span> <span class="op">&gt;</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span>, <span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        Pheno</span></span>
<span><span class="co">## Samp_1     0</span></span>
<span><span class="co">## Samp_2     1</span></span>
<span><span class="co">## Samp_3     0</span></span>
<span><span class="co">## Samp_4     1</span></span>
<span><span class="co">## Samp_5     1</span></span>
<span><span class="co">## Samp_6     0</span></span></code></pre>
<p>Denote the number of features in <span class="math inline">\(X_1
\&amp; X_2\)</span> as <span class="math inline">\(p_1 \&amp;
p_2\)</span> respectively, and the number of subjects as <span class="math inline">\(n\)</span>.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">X1</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">X2</span><span class="op">)</span></span>
<span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">X1</span><span class="op">)</span></span>
<span><span class="va">AbarLabel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">X1</span>, <span class="va">X2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Although SmCCNet does not require normality, it calculates the
Pearson correlation between linear combinations of omics features and
the phenotype, which assumes finite variances and finite covariance. It
is necessary to include a transformation if the data are skewed. The
algorithm also requires the data to be standardizable (i.e. none of the
data matrices include a column with zero variance.)</p>
</div>
<div class="section level3">
<h3 id="step-ii-determine-optimal-sparsity-penalties-through-cross-validation-optional-1">Step II: Determine optimal sparsity penalties through
cross-validation (optional)<a class="anchor" aria-label="anchor" href="#step-ii-determine-optimal-sparsity-penalties-through-cross-validation-optional-1"></a>
</h3>
<p>All the parameters set up are the same as multi-omics with
quantitative phenotype except that there is an extra penalty term to
evaluate: lasso penalty for SPLSDA classifier.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># number of folds in K-fold CV.</span></span>
<span><span class="va">K</span> <span class="op">&lt;-</span> <span class="fl">3</span> </span>
<span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">X1</span><span class="op">)</span></span>
<span><span class="co"># create a list of omics data</span></span>
<span><span class="va">X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">X1</span>, <span class="va">X2</span><span class="op">)</span></span>
<span><span class="co"># number of component for PLS</span></span>
<span><span class="va">ncomp</span> <span class="op">&lt;-</span> <span class="fl">3</span></span>
<span><span class="co"># number of omics dataset</span></span>
<span><span class="va">num_omics</span> <span class="op">&lt;-</span> <span class="fl">2</span></span>
<span><span class="co"># tuning parameter candidate length for each omics data</span></span>
<span><span class="va">tuneLength</span> <span class="op">&lt;-</span> <span class="fl">5</span></span>
<span><span class="co"># tuning parameter candadate range for each omics data</span></span>
<span><span class="va">minTune</span> <span class="op">&lt;-</span> <span class="fl">0.1</span></span>
<span><span class="va">maxTune</span> <span class="op">&lt;-</span> <span class="fl">0.5</span></span>
<span><span class="co"># create empty matrix to store all possible penalty parameters</span></span>
<span><span class="va">penSelect</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">0</span>, nrow <span class="op">=</span> <span class="va">tuneLength</span>, ncol <span class="op">=</span> <span class="va">num_omics</span><span class="op">)</span></span>
<span><span class="co"># set up the evaluation metric (choose between 'accuracy', 'auc', 'precision',</span></span>
<span><span class="co"># 'recall', 'f1')</span></span>
<span><span class="va">metric</span> <span class="op">&lt;-</span> <span class="st">'auc'</span></span>
<span><span class="co"># create sparsity penalty options.</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">Idx</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">penSelect</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">{</span></span>
<span>  <span class="va">penSelect</span><span class="op">[</span>,<span class="va">Idx</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span>from <span class="op">=</span> <span class="va">minTune</span>, </span>
<span>                         to <span class="op">=</span> <span class="va">maxTune</span>, </span>
<span>                         length.out <span class="op">=</span> <span class="va">tuneLength</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co"># combine with penalty term for classifier</span></span>
<span><span class="va">penSelect</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">penSelect</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fl">0.5</span>, </span>
<span>                         to <span class="op">=</span> <span class="fl">0.9</span>, </span>
<span>                         length.out <span class="op">=</span> <span class="va">tuneLength</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># expand grid</span></span>
<span><span class="co"># convert matrix to list of columns</span></span>
<span><span class="va">list_cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">as.list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">penSelect</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># generate all possible combinations</span></span>
<span><span class="va">PenExpand</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">expand.grid</span>, <span class="va">list_cols</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># set a CV directory.</span></span>
<span><span class="va">CVDir</span> <span class="op">&lt;-</span> <span class="st">"Example3foldCVBinary/"</span> </span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">CVDir</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="create-test-and-training-data-sets--1">Create test and training data sets.<a class="anchor" aria-label="anchor" href="#create-test-and-training-data-sets--1"></a>
</h4>
<p>Same as before, we need to split the data (<span class="math inline">\(X_1, X_2, Y\)</span>) into test and training sets
(Figure <span class="math inline">\(\ref{fig:CVflow}\)</span>, Step
I.1). Here we scale the complete dataset rather than each fold to
increase the performance in predicting the testing phenotype:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">12345</span><span class="op">)</span> <span class="co"># set random seed.</span></span>
<span></span>
<span><span class="co"># split data into folds</span></span>
<span><span class="va">X</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">X</span>, <span class="va">scale</span><span class="op">)</span></span>
<span><span class="va">foldIdx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">X</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">X</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, <span class="va">K</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">folddata</span> <span class="op">&lt;-</span> <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">foldIdx</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">Y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span></span>
<span>    <span class="va">X_train</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="va">X_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="va">Y_train</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="va">Y_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">{</span></span>
<span>      <span class="va">X_train</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">X</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="op">-</span><span class="va">foldIdx</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span>,<span class="op">]</span></span>
<span>      <span class="va">X_test</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">X</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">foldIdx</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span>,<span class="op">]</span></span>
<span>    <span class="op">}</span></span>
<span>    <span class="va">Y_train</span> <span class="op">&lt;-</span> <span class="va">Y</span><span class="op">[</span><span class="op">-</span><span class="va">foldIdx</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span>,<span class="op">]</span></span>
<span>    <span class="va">Y_test</span> <span class="op">&lt;-</span> <span class="va">Y</span><span class="op">[</span><span class="va">foldIdx</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span>,<span class="op">]</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>X_train <span class="op">=</span> <span class="va">X_train</span>, X_test <span class="op">=</span> <span class="va">X_test</span>,Y_train <span class="op">=</span> <span class="va">Y_train</span>, </span>
<span>                Y_test <span class="op">=</span> <span class="va">Y_test</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="co"># name each fold of data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">folddata</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'Fold_'</span>, <span class="fl">1</span><span class="op">:</span><span class="va">K</span><span class="op">)</span></span>
<span><span class="co"># saving all preliminary data into local directory for reproducibility purpose</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span><span class="va">folddata</span>, <span class="va">PenExpand</span>,</span>
<span>       file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">CVDir</span>, <span class="st">"CVData.RData"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="set-scaling-factors-1">Set Scaling Factors<a class="anchor" aria-label="anchor" href="#set-scaling-factors-1"></a>
</h4>
<p>It is important to set scaling factors for each pairwise correlation
to prioritize correlation structure(s) of interest. If two omics data
are used, there is only one scaling factor for between-omics canonical
correlation analysis, then it is natural to set the scaling factor to
1:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scalingFactor</span> <span class="op">&lt;-</span> <span class="fl">1</span></span></code></pre></div>
<p>If more than two omics data are used, then below is the function that
use prompt to help user define the scaling factor intended for the
analysis, note that different from the quantitative phenotype, the
phenotype should be excluded from the scaling factor input:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scalingFactor</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scalingFactorInput.html">scalingFactorInput</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'mRNA'</span>, <span class="st">'miRNA'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="run-k-fold-cv-1">Run K-fold CV<a class="anchor" aria-label="anchor" href="#run-k-fold-cv-1"></a>
</h4>
<p>Same as multi-omics with quantitative phenotype, cross-validation
should be run to determine the best penalty terms combinations. The
difference here is that since this is a predictive modeling problem, the
evaluation metric used should be based on the testing data, not the
comparison between training and testing. In the example here, we use the
testing AUC score to evaluate the model performance.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create an empty list to store the cv result </span></span>
<span><span class="va">CVResult</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># load cross-validation data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">CVDir</span>, <span class="st">"CVData.RData"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">CVidx</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">K</span><span class="op">)</span></span>
<span><span class="op">{</span>  </span>
<span>  <span class="va">CCcoef</span> <span class="op">&lt;-</span> <span class="va">scalingFactor</span></span>
<span>  <span class="va">TrainScore</span> <span class="op">&lt;-</span> <span class="va">TestScore</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">PenExpand</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="kw">for</span><span class="op">(</span><span class="va">idx</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">PenExpand</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="co"># consider one pair of sparsity penalties at a time.</span></span>
<span>    <span class="va">l</span> <span class="op">&lt;-</span> <span class="va">PenExpand</span><span class="op">[</span><span class="va">idx</span>, <span class="op">]</span></span>
<span>    <span class="co"># run multi-block PLS</span></span>
<span>    <span class="va">CCcoef</span> <span class="op">&lt;-</span> <span class="va">scalingFactor</span></span>
<span>    <span class="co"># run multi-block PLS</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="va">projection</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getRobustWeightsMultiBinary.html">getRobustWeightsMultiBinary</a></span><span class="op">(</span></span>
<span>                                   <span class="va">folddata</span><span class="op">[[</span><span class="va">CVidx</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"X_train"</span><span class="op">]</span><span class="op">]</span>, </span>
<span>                                   <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">folddata</span><span class="op">[[</span><span class="va">CVidx</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"Y_train"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, </span>
<span>                                   SubsamplingPercent<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>,</span>
<span>                                   Between_Discriminate_Ratio <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>,</span>
<span>                                   LambdaBetween <span class="op">=</span> <span class="va">l</span><span class="op">[</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">:</span><span class="va">num_omics</span><span class="op">]</span>, </span>
<span>                                   LambdaPheno <span class="op">=</span> <span class="va">l</span><span class="op">[</span><span class="fl">1</span>,<span class="op">(</span><span class="va">num_omics</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span><span class="op">]</span>, </span>
<span>                                   SubsamplingNum <span class="op">=</span> <span class="fl">1</span>, </span>
<span>                                   CCcoef <span class="op">=</span> <span class="va">CCcoef</span>,</span>
<span>                                   ncomp_pls <span class="op">=</span> <span class="va">ncomp</span>, EvalClassifier <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                   testData <span class="op">=</span> <span class="va">folddata</span><span class="op">[[</span><span class="va">CVidx</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"X_test"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>         </span>
<span>        </span>
<span></span>
<span>    <span class="co"># create training and testing data, and fit logistic regression model</span></span>
<span>    <span class="va">train_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">projection</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, </span>
<span>                             y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">folddata</span><span class="op">[[</span><span class="va">CVidx</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"Y_train"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">test_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>x <span class="op">=</span>  <span class="va">projection</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co"># catching error when performing the logistic regression</span></span>
<span>    <span class="va">has_error</span> <span class="op">&lt;-</span> <span class="cn">FALSE</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/conditions.html" class="external-link">tryCatch</a></span><span class="op">(</span><span class="op">{</span></span>
<span>        <span class="co"># fit logistic regression model</span></span>
<span>        <span class="va">logisticFit</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/glm.html" class="external-link">glm</a></span><span class="op">(</span><span class="va">y</span> <span class="op">~</span> <span class="va">.</span>, family <span class="op">=</span> <span class="st">'binomial'</span>, data <span class="op">=</span> <span class="va">train_data</span><span class="op">)</span></span>
<span>        <span class="co"># make prediction for train/test set</span></span>
<span>        <span class="va">train_pred</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">logisticFit</span>, <span class="va">train_data</span>, type <span class="op">=</span> <span class="st">'response'</span><span class="op">)</span></span>
<span>        <span class="va">test_pred</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">logisticFit</span>, <span class="va">test_data</span>, type <span class="op">=</span> <span class="st">'response'</span><span class="op">)</span></span>
<span>        <span class="va">train_score</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/classifierEval.html">classifierEval</a></span><span class="op">(</span>obs <span class="op">=</span> <span class="va">folddata</span><span class="op">[[</span><span class="va">CVidx</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"Y_train"</span><span class="op">]</span><span class="op">]</span>, </span>
<span>                                      pred <span class="op">=</span> <span class="va">train_pred</span>, </span>
<span>                                      EvalMethod <span class="op">=</span> <span class="va">metric</span>, print_score <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span>        <span class="va">test_score</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/classifierEval.html">classifierEval</a></span><span class="op">(</span>obs <span class="op">=</span> <span class="va">folddata</span><span class="op">[[</span><span class="va">CVidx</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"Y_test"</span><span class="op">]</span><span class="op">]</span>, </span>
<span>                                      pred <span class="op">=</span> <span class="va">test_pred</span>, </span>
<span>                                      EvalMethod <span class="op">=</span> <span class="va">metric</span>, print_score <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span>          <span class="op">}</span>,</span>
<span>        error <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">e</span><span class="op">)</span> <span class="op">{</span></span>
<span>          <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"Caught an error:"</span>, <span class="va">e</span><span class="op">$</span><span class="va">message</span>, <span class="st">"\n"</span><span class="op">)</span></span>
<span>          <span class="va">has_error</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span></span>
<span>        <span class="op">}</span><span class="op">)</span></span>
<span>        <span class="op">)</span></span>
<span>          </span>
<span>        <span class="va">TrainScore</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">train_score</span>, digits <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span>        <span class="va">TestScore</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">test_score</span>, digits <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span>    </span>
<span>    <span class="op">}</span></span>
<span>    </span>
<span>    <span class="co"># record prediction errors for given CV fold and all sparsity penalty </span></span>
<span>    <span class="co"># options.</span></span>
<span>    <span class="va">CVResult</span><span class="op">[[</span><span class="va">CVidx</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">TrainScore</span>, <span class="va">TestScore</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="extract-penalty-terms-with-the-highest-testing-evaluation-score">Extract penalty terms with the highest testing evaluation score<a class="anchor" aria-label="anchor" href="#extract-penalty-terms-with-the-highest-testing-evaluation-score"></a>
</h4>
<p>Finally, we extract the total prediction score and conclude the best
penalty pair as the pair with the highest prediction score (in the
example’s context, AUC score).</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># aggregate CV result and select the best penalty term</span></span>
<span><span class="va">AggregatedCVResult</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/funprog.html" class="external-link">Reduce</a></span><span class="op">(</span><span class="st">"+"</span>, <span class="va">CVResult</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">CVResult</span><span class="op">)</span></span>
<span><span class="co"># determine the best CV result</span></span>
<span><span class="va">optIdx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.min.html" class="external-link">which.max</a></span><span class="op">(</span><span class="va">AggregatedCVResult</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>In the last step, we store the aggregated cross-validation result
into the local directory:</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># combine CV evaluation result with penalty candidates </span></span>
<span><span class="va">overallCVInfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">PenExpand</span>, <span class="va">AggregatedCVResult</span><span class="op">)</span></span>
<span><span class="co"># set column names for penalty terms for omics</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">overallCVInfo</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">num_omics</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'l'</span>,<span class="fl">1</span><span class="op">:</span><span class="va">num_omics</span><span class="op">)</span></span>
<span><span class="co"># set column names for penalty terms for classifier</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">overallCVInfo</span><span class="op">)</span><span class="op">[</span><span class="va">num_omics</span><span class="op">+</span><span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'lpheno'</span><span class="op">)</span></span>
<span><span class="co"># save overall CV result </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.csv</a></span><span class="op">(</span><span class="va">overallCVInfo</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">CVDir</span>, <span class="st">'overallCVInfo.csv'</span><span class="op">)</span>, </span>
<span>          row.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co"># print out the best CV penalty pair and associated result</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">overallCVInfo</span><span class="op">[</span><span class="va">optIdx</span>,<span class="op">]</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="step-iii-run-splsda-with-pre-selected-penalty-terms">Step III: Run SPLSDA with pre-selected penalty terms<a class="anchor" aria-label="anchor" href="#step-iii-run-splsda-with-pre-selected-penalty-terms"></a>
</h3>
<p>Same as above, with a pre-selected penalty pair, we apply our method
to subsampled features of <span class="math inline">\(X_1, X_2\)</span>
and <span class="math inline">\(Y\)</span>, and repeat the process to
generate a robust similarity matrix. Note that in addition to the
subsampling, we need to add a between-discriminant ratio (defined as
<span class="math inline">\(\gamma_1/\gamma_2\)</span>) to control for
the trade-off between omics-omics relationship and omics-phenotype
relationship presented in the subsequent adjacency matrix. The larger
the second number is, the more emphasize is on the omics-phenotype
correlation, and vice versa.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># feature sampling proportions, 0.9 for miRNA since it has less features.</span></span>
<span><span class="va">s</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.7</span>,<span class="fl">0.9</span><span class="op">)</span> </span>
<span><span class="co"># number of subsamples.</span></span>
<span><span class="va">SubsamplingNum</span> <span class="op">&lt;-</span> <span class="fl">50</span></span>
<span><span class="co"># run SPLSDA on the subsamples</span></span>
<span><span class="va">Ws</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getRobustWeightsMultiBinary.html">getRobustWeightsMultiBinary</a></span><span class="op">(</span><span class="va">X</span>, </span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span>, </span>
<span>            SubsamplingPercent<span class="op">=</span><span class="va">s</span>,</span>
<span>            Between_Discriminate_Ratio <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>,</span>
<span>            LambdaBetween <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">overallCVInfo</span><span class="op">[</span><span class="va">optIdx</span>,<span class="fl">1</span><span class="op">:</span><span class="va">num_omics</span><span class="op">]</span><span class="op">)</span>, </span>
<span>            LambdaPheno <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">overallCVInfo</span><span class="op">[</span><span class="va">optIdx</span>,<span class="va">num_omics</span> <span class="op">+</span> <span class="fl">1</span><span class="op">]</span><span class="op">)</span>, </span>
<span>            SubsamplingNum <span class="op">=</span> <span class="va">SubsamplingNum</span>, </span>
<span>            CCcoef <span class="op">=</span> <span class="va">scalingFactor</span>,</span>
<span>            ncomp_pls <span class="op">=</span> <span class="va">ncomp</span>, EvalClassifier <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>The rest of the network analysis step is the same as the multi-omics
SmCCNet with quantitative phenotype, please refer to section 2.3 and 2.4
for further information. Note that since the cross-validation penalty
terms evaluation with binary phenotype doesn’t involve the participation
of canonical weight from step 1, there is no need to tune the scaling
factors with the method 3 above in section 4.3. However, method 1 and 2
(section 4.1 and 4.2) can be used to determine the scaling factors for
between-omics relationship if there are more than 2 omics data.</p>
</div>
</div>
<div class="section level2">
<h2 id="acknowledgments">Acknowledgments<a class="anchor" aria-label="anchor" href="#acknowledgments"></a>
</h2>
<p>We would like to acknowledge the original contributors to the SmCCNet
R package, including Yonghua Zhuang, W. Jenny Shi, and Laura Saba. Their
collective contributions to the earlier versions of SmCCNet have been
instrumental in shaping its development.</p>
<p>We are also deeply grateful to Dr.Katerina Kechris, for her
continuous guidance and support, which has been vital in the refinement
of this package.</p>
<p>The SmCCNet R package owes much of its growth and functionality to
these original contributors, and we appreciate their enduring influence
on its development.</p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.2 (2023-10-31)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] grid      stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] reshape2_1.4.4   shadowtext_0.1.2 lubridate_1.9.3  forcats_1.0.0   </span></span>
<span><span class="co">##  [5] stringr_1.5.1    dplyr_1.1.4      purrr_1.0.2      readr_2.1.5     </span></span>
<span><span class="co">##  [9] tidyr_1.3.0      tibble_3.2.1     ggplot2_3.4.4    tidyverse_2.0.0 </span></span>
<span><span class="co">## [13] SmCCNet_2.0.2    furrr_0.3.1      future_1.33.1    igraph_1.6.0    </span></span>
<span><span class="co">## [17] Matrix_1.6-1.1   pbapply_1.7-2   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] gtable_0.3.4      xfun_0.41         bslib_0.6.1       lattice_0.21-9   </span></span>
<span><span class="co">##  [5] tzdb_0.4.0        vctrs_0.6.5       tools_4.3.2       generics_0.1.3   </span></span>
<span><span class="co">##  [9] parallel_4.3.2    fansi_1.0.6       highr_0.10        pkgconfig_2.0.3  </span></span>
<span><span class="co">## [13] desc_1.4.3        lifecycle_1.0.4   farver_2.1.1      compiler_4.3.2   </span></span>
<span><span class="co">## [17] textshaping_0.3.7 munsell_0.5.0     codetools_0.2-19  htmltools_0.5.7  </span></span>
<span><span class="co">## [21] sass_0.4.8        yaml_2.3.8        pillar_1.9.0      pkgdown_2.0.7    </span></span>
<span><span class="co">## [25] jquerylib_0.1.4   cachem_1.0.8      parallelly_1.36.0 tidyselect_1.2.0 </span></span>
<span><span class="co">## [29] digest_0.6.34     stringi_1.8.3     listenv_0.9.0     labeling_0.4.3   </span></span>
<span><span class="co">## [33] fastmap_1.1.1     colorspace_2.1-0  cli_3.6.2         magrittr_2.0.3   </span></span>
<span><span class="co">## [37] utf8_1.2.4        withr_2.5.2       scales_1.3.0      timechange_0.2.0 </span></span>
<span><span class="co">## [41] rmarkdown_2.25    globals_0.16.2    ragg_1.2.7        hms_1.1.3        </span></span>
<span><span class="co">## [45] memoise_2.0.1     evaluate_0.23     knitr_1.45        rlang_1.1.3      </span></span>
<span><span class="co">## [49] Rcpp_1.0.12       glue_1.7.0        jsonlite_1.8.8    R6_2.5.1         </span></span>
<span><span class="co">## [53] plyr_1.8.9        systemfonts_1.0.5 fs_1.6.3</span></span></code></pre>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/warnings.html" class="external-link">warnings</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<blockquote>
<p>Chung D and Keles S (2010), “Sparse partial least squares
classification for high dimensional data”, Statistical Applications in
Genetics and Molecular Biology, Vol. 9, Article 17.</p>
</blockquote>
<blockquote>
<p>Shi, W.J., Y. Zhuang, P.H. Russell, B.D. Hobbs, M.M. Parker,P.J.
Castaldi, P. Rudra, B. Vestal, C.P. Hersh, L.M. Saba, and K. Kechris,
“Unsupervised Discovery of Phenotype Specific Multi-Omics Networks.”
Bioinformatics. 2019 Nov 1;35(21):4336-4343. doi:
10.1093/bioinformatics/btz226.</p>
</blockquote>
<blockquote>
<p>Shannon, P., Markiel, A., Ozier, O., Baliga, N. S., Wang, J. T.,
Ramage, D., … &amp; Ideker, T. (2003). Cytoscape: a software environment
for integrated models of biomolecular interaction networks. Genome
research, 13(11), 2498-2504.</p>
</blockquote>
<blockquote>
<p>Vu, T., Litkowski, E. M., Liu, W., Pratte, K. A., Lange, L., Bowler,
R. P., … &amp; Kechris, K. J. (2023). NetSHy: network summarization via
a hybrid approach leveraging topological properties. Bioinformatics,
39(1), btac818.</p>
</blockquote>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Weixuan Liu.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
