<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="SmCCNet">
<title>Reconstructing Phenotype-Specific Single-Omics Networks with SmCCNet â€¢ SmCCNet</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Reconstructing Phenotype-Specific Single-Omics Networks with SmCCNet">
<meta property="og:description" content="SmCCNet">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">SmCCNet</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/SmCCNet_Vignette_AutoSmCCNet.html">Automated SmCCNet</a>
    <a class="dropdown-item" href="../articles/SmCCNet_Vignette_MultiOmics.html">Reconstructing Phenotype-Specific Multi-Omics Networks with SmCCNet</a>
    <a class="dropdown-item" href="../articles/SmCCNet_Vignette_SingleOmics.html">Reconstructing Phenotype-Specific Single-Omics Networks with SmCCNet</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/KechrisLab/SmCCNet/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Reconstructing Phenotype-Specific Single-Omics Networks with SmCCNet</h1>
                        <h4 data-toc-skip class="author">Weixuan
Liu</h4>
                        <h4 data-toc-skip class="author">Katerina
Kechris</h4>
            
            <h4 data-toc-skip class="date">2024-01-14</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/KechrisLab/SmCCNet/blob/HEAD/vignettes/SmCCNet_Vignette_SingleOmics.Rmd" class="external-link"><code>vignettes/SmCCNet_Vignette_SingleOmics.Rmd</code></a></small>
      <div class="d-none name"><code>SmCCNet_Vignette_SingleOmics.Rmd</code></div>
    </div>

    
    
<div class="section level3">
<h3 id="smccnet-package">SmCCNet package<a class="anchor" aria-label="anchor" href="#smccnet-package"></a>
</h3>
<p>The SmCCNet package has the following dependencies:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/psolymos/pbapply" class="external-link">pbapply</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://Matrix.R-forge.R-project.org" class="external-link">Matrix</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r.igraph.org/" class="external-link">igraph</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/KechrisLab/SmCCNet" class="external-link">SmCCNet</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">parallel</span><span class="op">)</span></span></code></pre></div>
<p>This section provides a tutorial on running the SmCCNet algorithm for
a quantitative phenotype using single-omics data. Please see other
vignettes for multi-omics applications.</p>
</div>
<div class="section level2">
<h2 id="smccnet-single-omics-workflow-with-a-synthetic-dataset">SmCCNet single-omics workflow with a synthetic dataset<a class="anchor" aria-label="anchor" href="#smccnet-single-omics-workflow-with-a-synthetic-dataset"></a>
</h2>
<div class="figure">
<img src="figures/single-omics-smccnet.jpg" alt="Workflow for SmCCNet single-omics setting (binary and quantitative phenotype)." width="100%"><p class="caption">
Workflow for SmCCNet single-omics setting (binary and quantitative
phenotype).
</p>
</div>
<p>As an example, we consider a synthetic data set with 500 genes <span class="math inline">\((X_1)\)</span> expression levels measured for 358
subjects, along with a quantitative phenotype <span class="math inline">\((Y)\)</span>. The example has a single phenotype,
but a multi-dimensional phenotype matrix is also possible.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">ExampleData</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">X1</span><span class="op">[</span> , <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          Gene_1   Gene_2   Gene_3   Gene_4   Gene_5   Gene_6</span></span>
<span><span class="co">## Samp_1 22.48570 40.35372 31.02575 20.84721 26.69729 30.20545</span></span>
<span><span class="co">## Samp_2 37.05885 34.05223 33.48702 23.53146 26.75463 31.73594</span></span>
<span><span class="co">## Samp_3 20.53077 31.66962 35.18957 20.95254 25.01883 32.15723</span></span>
<span><span class="co">## Samp_4 33.18689 38.48088 18.89710 31.82330 34.04938 38.79989</span></span>
<span><span class="co">## Samp_5 28.96198 41.06049 28.49496 18.37449 30.81524 24.00454</span></span>
<span><span class="co">## Samp_6 18.05983 29.55471 32.54002 29.68452 26.19996 26.76684</span></span></code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##           Pheno</span></span>
<span><span class="co">## Samp_1 235.0674</span></span>
<span><span class="co">## Samp_2 253.5450</span></span>
<span><span class="co">## Samp_3 234.2050</span></span>
<span><span class="co">## Samp_4 281.0354</span></span>
<span><span class="co">## Samp_5 245.4478</span></span>
<span><span class="co">## Samp_6 189.6231</span></span></code></pre>
<p>Denote the number of features in <span class="math inline">\(X_1\)</span> as <span class="math inline">\(p_1\)</span> and the number of subjects as <span class="math inline">\(n\)</span>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">X1</span><span class="op">)</span></span>
<span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">X1</span><span class="op">)</span></span>
<span><span class="va">AbarLabel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">X1</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="step-i--preprocessing">Step I. Preprocessing<a class="anchor" aria-label="anchor" href="#step-i--preprocessing"></a>
</h3>
<p>The first step is to preprocess the single omics data. All canonical
correlation analysis (CCA) methods require data sets to be standardized
(centered and scaled) by columns (e.g.Â features) to ensure the
equivalent contribution of each feature when maximizing covariance. In
addition, there are some other optional steps to preprocess the data.
Note that when employing both standardization and covariate adjustment,
it is essential to apply standardization prior to covariate adjustment.
This sequencing ensures our best efforts to meet the assumptions of
linear regression In our data preprocessing pipeline, the following
options are given (with ordering):</p>
<ul>
<li>
<strong>1. CoV Filtering</strong>: Filter features based on
coefficients of variation (CoV).</li>
<li>
<strong>2. Standardization</strong>: Center and/or scale data.</li>
<li>
<strong>3. Adjust for Covariates</strong>: Regress out specified
covariates and return residuals</li>
</ul>
<p>Below is an example of data preprocessing with only feature filtering
and standardization, in this case, there is no covariate adjustment, the
coeffcient of variation filtering quantile is 0.2 (meaning that features
with CoV smaller than 20% quantile of CoV will be filtered out), and
data are centered and scaled:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># preprocess data</span></span>
<span><span class="va">processed_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dataPreprocess.html">dataPreprocess</a></span><span class="op">(</span>X <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">X1</span><span class="op">)</span>, covariates <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>                  is_cv <span class="op">=</span> <span class="cn">TRUE</span>, cv_quantile <span class="op">=</span> <span class="fl">0.2</span>, center <span class="op">=</span> <span class="cn">TRUE</span>, scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="step-ii-determine-optimal-sparsity-penalties-through-cross-validation-optional">Step II: Determine optimal sparsity penalties through
cross-validation (optional)<a class="anchor" aria-label="anchor" href="#step-ii-determine-optimal-sparsity-penalties-through-cross-validation-optional"></a>
</h3>
<p>To find the optimal sparsity penalties, we apply a K-fold
cross-validation on the synthetic data (Figure 1). Let <span class="math inline">\(p_1\)</span> denote the number of features in
omics data <span class="math inline">\(X_1\)</span> respectively, and
<span class="math inline">\(s_1\)</span> is the proportion of <span class="math inline">\(X_1\)</span> features to be sampled every time.
The sparse penalties range from 0.05 to 0.3 with step size of 0.05.
Large penalty values correspond to less sparse canonical weight vector,
while small penalties correspond to sparser canonical weight vector.
Below is the list of parameters that need to be specified:</p>
<ul>
<li>
<span class="math inline">\(K\)</span>: Number of folds in
cross-validation (CV). Typically a 5-fold CV is sufficient. If the
training set contains too few (e.g.Â <span class="math inline">\(&lt;30\)</span>) samples, or the test or training
set becomes unscalable, then choose a smaller <span class="math inline">\(K\)</span>.</li>
</ul>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">K</span> <span class="op">&lt;-</span> <span class="fl">3</span> <span class="co"># number of folds in K-fold CV.</span></span>
<span><span class="va">s1</span> <span class="op">&lt;-</span> <span class="fl">0.7</span></span>
<span><span class="va">subSamp</span> <span class="op">&lt;-</span> <span class="fl">50</span> <span class="co"># number of subsamples (will be described in later section).</span></span>
<span></span>
<span><span class="co"># create sparsity penalty options.</span></span>
<span><span class="va">pen1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">.05</span>, <span class="fl">.3</span>, by <span class="op">=</span> <span class="fl">.05</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># set a CV directory.</span></span>
<span><span class="va">CVDir</span> <span class="op">&lt;-</span> <span class="st">"Example3foldCV/"</span> </span>
<span><span class="va">pheno</span> <span class="op">&lt;-</span> <span class="st">"Example3foldCV"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">CVDir</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="create-test-and-training-data-sets-">Create test and training data sets.<a class="anchor" aria-label="anchor" href="#create-test-and-training-data-sets-"></a>
</h4>
<p>To perform K-fold Cross-Validation (CV), we need to split the data
(<span class="math inline">\(X_1, Y\)</span>) into test and training
sets. We have included the standardization step within the SmCCNet
algorithm. However, for the CV procedure, we recommend to standardize
the training and test sets upfront, since this helps to choose the
number of CV folds <span class="math inline">\(K\)</span>. If any data
set can not be standardized, we recommend to reduce <span class="math inline">\(K\)</span>. In the code below, we show how to
create CV data sets and check if all data sets are valid
(i.e.Â standardizable). The standardized training and test data sets will
be saved under the CV directory.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># set random seed</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">12345</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># save data and parameters into local directory</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span><span class="va">X1</span>, <span class="va">Y</span>, <span class="va">s1</span>, <span class="va">subSamp</span>, <span class="va">pen1</span>, </span>
<span>     file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">CVDir</span>, <span class="st">"Data.Rdata"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># split data into K folds</span></span>
<span><span class="va">foldIdx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">N</span>, <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">N</span>, <span class="va">K</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">K</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">iIdx</span> <span class="op">&lt;-</span> <span class="va">foldIdx</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="va">x1.train</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/scale.html" class="external-link">scale</a></span><span class="op">(</span><span class="va">X1</span><span class="op">[</span><span class="op">-</span><span class="va">iIdx</span>, <span class="op">]</span><span class="op">)</span></span>
<span>  <span class="va">yy.train</span> <span class="op">&lt;-</span> <span class="va">Y</span><span class="op">[</span><span class="op">-</span><span class="va">iIdx</span>, <span class="op">]</span></span>
<span>  <span class="va">x1.test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/scale.html" class="external-link">scale</a></span><span class="op">(</span><span class="va">X1</span><span class="op">[</span><span class="va">iIdx</span>, <span class="op">]</span><span class="op">)</span></span>
<span>  <span class="va">yy.test</span> <span class="op">&lt;-</span> <span class="va">Y</span><span class="op">[</span><span class="va">iIdx</span>, <span class="op">]</span></span>
<span>  </span>
<span>  <span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">x1.train</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">yy.train</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">x1.test</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">yy.test</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/stop.html" class="external-link">stop</a></span><span class="op">(</span><span class="st">"Invalid scaled data."</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>  <span class="va">subD</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">CVDir</span>, <span class="st">"CV_"</span>, <span class="va">i</span>, <span class="st">"/"</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">subD</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># save data from each fold into local directory</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span><span class="va">x1.train</span>, <span class="va">yy.train</span>, <span class="va">x1.test</span>, <span class="va">yy.test</span>,</span>
<span>       <span class="va">pen1</span>, <span class="va">p1</span>, </span>
<span>       file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">subD</span>, <span class="st">"Data.Rdata"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="run-k-fold-cross-validation">Run K-fold Cross-Validation<a class="anchor" aria-label="anchor" href="#run-k-fold-cross-validation"></a>
</h4>
<p>For each of the K-folds, we compute the prediction error for each
penalty pair option. Since there is no subsampling step for
cross-validation, we run through cross-validation with nested for loop.
However, if the omics data are extremely high-dimensional, we recommend
using the R package <strong>parallel</strong> to parallelize the for
loop, or use <strong>fastAutoSmCCNet()</strong> directly.
<strong>fastAutoSmCCNet()</strong> is the package built-in function that
streamline the pipeline with single line of code, and the
cross-validation step is parallelized with the
<strong>future_map()</strong> in <strong>furrr</strong> package.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># number of clusters in parSapply should be the same as number specified above</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="kw">for</span> <span class="op">(</span><span class="va">CVidx</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">K</span><span class="op">)</span></span>
<span><span class="op">{</span></span>
<span>  <span class="co"># define the sub-directory for each fold </span></span>
<span>  <span class="va">subD</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">CVDir</span>, <span class="st">"CV_"</span>, <span class="va">CVidx</span>, <span class="st">"/"</span><span class="op">)</span></span>
<span>  <span class="co"># load fold data</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">subD</span>, <span class="st">"Data.Rdata"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">subD</span>, <span class="st">"SmCCA/"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="co"># create empty vector to store cross-validation result</span></span>
<span>  <span class="va">RhoTrain</span> <span class="op">&lt;-</span> <span class="va">RhoTest</span> <span class="op">&lt;-</span> <span class="va">DeltaCor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">pen1</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="co"># evaluate through all the possible penalty candidates</span></span>
<span>  <span class="kw">for</span><span class="op">(</span><span class="va">idx</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">pen1</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">l1</span> <span class="op">&lt;-</span> <span class="va">pen1</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Running SmCCA on CV_"</span>, <span class="va">CVidx</span>, <span class="st">" pen="</span>, <span class="va">l1</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="co"># run single-omics SmCCNet</span></span>
<span>    <span class="va">Ws</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getRobustWeightsSingle.html">getRobustWeightsSingle</a></span><span class="op">(</span><span class="va">x1.train</span>, <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">yy.train</span><span class="op">)</span>, <span class="va">l1</span>, <span class="fl">1</span>,</span>
<span>                                    SubsamplingNum <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span>    <span class="co"># average</span></span>
<span>    <span class="va">meanW</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="va">Ws</span><span class="op">)</span></span>
<span>    <span class="va">v</span> <span class="op">&lt;-</span> <span class="va">meanW</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">p1</span><span class="op">]</span></span>
<span>    </span>
<span>    <span class="va">rho.train</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">x1.train</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span> <span class="va">v</span>, <span class="va">yy.train</span><span class="op">)</span></span>
<span>    </span>
<span>    </span>
<span>    <span class="va">rho.test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">x1.test</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span> <span class="va">v</span>, <span class="va">yy.test</span><span class="op">)</span> </span>
<span>    </span>
<span>    </span>
<span>    <span class="va">RhoTrain</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">rho.train</span>, digits <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span>    <span class="va">RhoTest</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">rho.test</span>, digits <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span>    <span class="va">DeltaCor</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">rho.train</span> <span class="op">-</span> <span class="va">rho.test</span><span class="op">)</span></span>
<span>    </span>
<span>   </span>
<span>    </span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>  <span class="va">DeltaCor.all</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">pen1</span>, <span class="va">RhoTrain</span>, <span class="va">RhoTest</span>, <span class="va">DeltaCor</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">DeltaCor.all</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"l1"</span>, <span class="st">"Training CC"</span>, <span class="st">"Test CC"</span>, <span class="st">"CC Pred. Error"</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.csv</a></span><span class="op">(</span><span class="va">DeltaCor.all</span>,</span>
<span>            file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">subD</span>, <span class="st">"SmCCA/SCCA_"</span>, <span class="va">subSamp</span>,<span class="st">"_allDeltaCor.csv"</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span>  </span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="extract-penalty-term-with-the-smallest-total-prediction-error">Extract penalty term with the smallest total prediction error<a class="anchor" aria-label="anchor" href="#extract-penalty-term-with-the-smallest-total-prediction-error"></a>
</h4>
<p>We extract the total prediction errors and return the best penalty
term. This step will automatically give the optimal testing canonical
correlation choice as well as prediction error choice, which serves as a
reference for the penalty term selection. There are different ways to
select the best penalty terms, one of the simplest way is to minimize
discrepancy between the training canonical correlation and the testing
canonical correlation. However, this method does not take the magnitude
of testing canonical correlation into account, which means it may end up
selecting the penalty term with smaller canonical correlation (low
signal). For instance, if a certain penalty term yields the training
canonical correlation of 0.7, with the testing canonical correlation of
0.4, and another penalty term yield the training canonical correlation
of 0.4, with the testing canonical correlation of 0.2, minimizing
training and testing canonical correlation selects the latter.
Therefore, in this step, we want to minimized the scaled prediction
error, which is defined as:</p>
<p><span class="math display">\[\begin{equation}
scaledPredErr = \frac{|trainCC - testCC|}{|testCC|},
\end{equation}\]</span></p>
<p>where <span class="math inline">\(trainCC\)</span> and <span class="math inline">\(testCC\)</span> is defined as the training
canonical correlation and testing canonical correlation
respectively.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># combine prediction errors from all K folds and compute the total prediction</span></span>
<span><span class="co"># error for each sparsity penalty pair.</span></span>
<span><span class="fu"><a href="../reference/aggregateCVSingle.html">aggregateCVSingle</a></span><span class="op">(</span><span class="va">CVDir</span>, <span class="st">"SmCCA"</span>, NumSubsamp <span class="op">=</span> <span class="va">subSamp</span>, K <span class="op">=</span> <span class="va">K</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="step-iii-run-smcca-with-pre-selected-penalty-term">Step III: Run SmCCA with pre-selected penalty term<a class="anchor" aria-label="anchor" href="#step-iii-run-smcca-with-pre-selected-penalty-term"></a>
</h3>
<p>With a pre-selected penalty term, we apply SmCCA to subsampled
features of <span class="math inline">\(X_1\)</span> and <span class="math inline">\(Y\)</span>, and repeat the process to generate a
robust similarity matrix (Figure 1). As for the number of subsample, a
larger number of subsamples leads to more stable results, while a
smaller number of subsample is faster computationally. We use 50 in this
example. Below is the setup and description of the subsampling
parameters:</p>
<ul>
<li>
<span class="math inline">\(s1\)</span>: Proportions of feature
subsampling from <span class="math inline">\(X_1, X_2\)</span>. Default
values are <span class="math inline">\(s_1 = 0.7, s_2 =
0.9.\)</span>
</li>
<li>
<span class="math inline">\(SubsamplingNum\)</span>: Number of
subsamples.</li>
</ul>
<p>After obtaining the canonical weight <span class="math inline">\(Ws\)</span>, which has the dimension of <span class="math inline">\(p_1\)</span> (number of features) by <span class="math inline">\(p_1 s_1\)</span> (proportion of feature
subsamples), the next step is to obtain the adjacency matrix by taking
the outer product of each <span class="math inline">\(W\)</span> with
itself to obtain an adjacency matrix and average the matrices to obtain
<span class="math inline">\(Abar\)</span>, a sparse matrix object.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># set up directory to store all the results</span></span>
<span><span class="va">plotD</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">CVDir</span>, <span class="st">"Figures/"</span><span class="op">)</span></span>
<span><span class="va">saveD</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">CVDir</span>, <span class="st">"Results/"</span><span class="op">)</span></span>
<span><span class="va">dataF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">CVDir</span>, <span class="st">"Data.Rdata"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">plotD</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">saveD</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">dataF</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># type of CCA result, only "SmCCA" supported</span></span>
<span><span class="va">Method</span> <span class="op">=</span> <span class="st">"SmCCA"</span></span>
<span></span>
<span></span>
<span><span class="co"># after SmCCA CV, select the best penalty term, </span></span>
<span><span class="co"># and use it for running SmCCA on the complete dataset</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">Method</span> <span class="kw">in</span> <span class="st">"SmCCA"</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="co"># select optimal penalty term from CV result</span></span>
<span>  <span class="va">T12</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">CVDir</span>, <span class="st">"Results/"</span>, <span class="va">Method</span>, <span class="st">"CVmeanDeltaCors.csv"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="co"># calculate evaluation metric **</span></span>
<span>  <span class="va">pen</span> <span class="op">&lt;-</span> <span class="va">T12</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.min.html" class="external-link">which.min</a></span><span class="op">(</span><span class="va">T12</span><span class="op">[</span> ,<span class="fl">4</span><span class="op">]</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">T12</span><span class="op">[</span> ,<span class="fl">3</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> ,<span class="fl">2</span><span class="op">]</span></span>
<span>  </span>
<span>  </span>
<span>  <span class="va">l1</span> <span class="op">&lt;-</span> <span class="va">pen</span>;</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="va">Ws</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getRobustWeightsSingle.html">getRobustWeightsSingle</a></span><span class="op">(</span>X1 <span class="op">=</span> <span class="va">X1</span>, Trait <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span>, </span>
<span>                                       Lambda1 <span class="op">=</span> <span class="va">l1</span>, </span>
<span>                                       <span class="va">s1</span>, SubsamplingNum <span class="op">=</span> <span class="va">subSamp</span><span class="op">)</span></span>
<span>    </span>
<span>    </span>
<span>    <span class="va">Abar</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getAbar.html">getAbar</a></span><span class="op">(</span><span class="va">Ws</span>, FeatureLabel <span class="op">=</span> <span class="va">AbarLabel</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">p1</span><span class="op">]</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span><span class="va">l1</span>, <span class="va">X1</span>, <span class="va">Y</span>, <span class="va">s1</span>, <span class="va">Ws</span>, <span class="va">Abar</span>,</span>
<span>         file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">saveD</span>, <span class="va">Method</span>, <span class="va">K</span>, <span class="st">"foldSamp"</span>, <span class="va">subSamp</span>, <span class="st">"_"</span>, <span class="va">pen</span>, </span>
<span>                       <span class="st">".Rdata"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span><span class="op">)</span></span>
<span>  </span>
<span>  </span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="step-iv-obtain-single-omics-modules-through-network-clustering">Step IV: Obtain single-omics modules through network clustering<a class="anchor" aria-label="anchor" href="#step-iv-obtain-single-omics-modules-through-network-clustering"></a>
</h3>
<p>From the similarity matrix obtained in the last step, we obtain
single-omics modules by applying hierarchical tree cutting and plotting
the reconstructed networks. The edge signs are recovered from pairwise
feature correlations.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># perform clustering based on the adjacency matrix Abar </span></span>
<span><span class="va">OmicsModule</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getOmicsModules.html">getOmicsModules</a></span><span class="op">(</span><span class="va">Abar</span>, PlotTree <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># make sure there are no duplicated labels</span></span>
<span><span class="va">AbarLabel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/make.unique.html" class="external-link">make.unique</a></span><span class="op">(</span><span class="va">AbarLabel</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># calculate feature correlation matrix</span></span>
<span><span class="va">bigCor2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">X1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># data type</span></span>
<span><span class="va">types</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">'gene'</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">bigCor2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="step-v-obtain-network-summarization-score-and-pruned-subnetworks">Step V: Obtain network summarization score and pruned
subnetworks<a class="anchor" aria-label="anchor" href="#step-v-obtain-network-summarization-score-and-pruned-subnetworks"></a>
</h3>
<p>The next step is to prune the network so that the unnecessary
features (nodes) will be filtered from the original network module. This
principle is based on a subject-level score of interest known as the
network summarization score. There are two different network
summarization methods: PCA and NetSHy (network summarization via a
hybrid approach, Vu et al 2023 Bioinformatics), which are specified by
the argument â€˜methodâ€™. We evaluate two criteria stepwise 1)
summarization score correlation with respect to the phenotype, which is
used to verify if the summarization score for the current subnetwork has
a strong signal with respect to the phenotype and 2) The correlation
between the summarization of the current subnetwork and that of the
baseline network with a pre-defined baseline network size. This is used
to check if the addition of more molecular features introduces noise.
The stepwise approach for network pruning is:</p>
<p>Users can prune each network module obtained through hierarchical
clustering with the following code:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># filter out network modules with insufficient number of nodes</span></span>
<span><span class="va">module_length</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">OmicsModule</span>, <span class="va">length</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">network_modules</span> <span class="op">&lt;-</span> <span class="va">OmicsModule</span><span class="op">[</span><span class="va">module_length</span> <span class="op">&gt;</span> <span class="fl">10</span><span class="op">]</span></span>
<span><span class="co"># extract pruned network modules</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">network_modules</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'For network module: '</span>, <span class="va">i</span>, <span class="st">'\n'</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="co"># define subnetwork</span></span>
<span>  <span class="va">abar_sub</span> <span class="op">&lt;-</span> <span class="va">Abar</span><span class="op">[</span><span class="va">network_modules</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>,<span class="va">network_modules</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="va">cor_sub</span> <span class="op">&lt;-</span> <span class="va">bigCor2</span><span class="op">[</span><span class="va">network_modules</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>,<span class="va">network_modules</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="co"># prune network module</span></span>
<span>  <span class="fu"><a href="../reference/networkPruning.html">networkPruning</a></span><span class="op">(</span>Abar <span class="op">=</span> <span class="va">abar_sub</span>,CorrMatrix <span class="op">=</span> <span class="va">cor_sub</span>, </span>
<span>                          type <span class="op">=</span> <span class="va">types</span><span class="op">[</span><span class="va">network_modules</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">]</span>, </span>
<span>                 data <span class="op">=</span> <span class="va">X1</span><span class="op">[</span>,<span class="va">network_modules</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">]</span>,      </span>
<span>              Pheno <span class="op">=</span> <span class="va">Y</span>, ModuleIdx <span class="op">=</span> <span class="va">i</span>, min_mod_size <span class="op">=</span> <span class="fl">10</span>, </span>
<span>                          max_mod_size <span class="op">=</span> <span class="fl">100</span>, method <span class="op">=</span> <span class="st">'PCA'</span>, </span>
<span>                          saving_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"\n"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>The output from this step contains a network adjacency matrix,
summarization scores (first 3 PCs), PC loadings and more, which are
stored in a .Rdata file in the user-specified location.</p>
</div>
</div>
<div class="section level2">
<h2 id="results">Results<a class="anchor" aria-label="anchor" href="#results"></a>
</h2>
<p>We present the single-omics network result based on the synthetic
data. The first table below contains the individual molecular features
correlation with respect to phenotype, and their associated p-value
(from correlation testing).</p>
<table class="table">
<caption>Individual molecular features correlation table with respect to
phenotype (correlation and p-value).</caption>
<thead><tr class="header">
<th align="left">Molecular Feature</th>
<th align="right">Correlation to Phenotype</th>
<th align="right">P-value</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Gene_1</td>
<td align="right">0.3828730</td>
<td align="right">0.0000000</td>
</tr>
<tr class="even">
<td align="left">Gene_2</td>
<td align="right">0.3411378</td>
<td align="right">0.0000000</td>
</tr>
<tr class="odd">
<td align="left">Gene_5</td>
<td align="right">0.1310036</td>
<td align="right">0.0131111</td>
</tr>
<tr class="even">
<td align="left">Gene_6</td>
<td align="right">0.6284530</td>
<td align="right">0.0000000</td>
</tr>
<tr class="odd">
<td align="left">Gene_7</td>
<td align="right">0.6531262</td>
<td align="right">0.0000000</td>
</tr>
<tr class="even">
<td align="left">Gene_28</td>
<td align="right">0.1306018</td>
<td align="right">0.0133961</td>
</tr>
<tr class="odd">
<td align="left">Gene_46</td>
<td align="right">0.1204872</td>
<td align="right">0.0226039</td>
</tr>
<tr class="even">
<td align="left">Gene_54</td>
<td align="right">-0.1385489</td>
<td align="right">0.0086656</td>
</tr>
<tr class="odd">
<td align="left">Gene_72</td>
<td align="right">-0.1274960</td>
<td align="right">0.0157889</td>
</tr>
<tr class="even">
<td align="left">Gene_74</td>
<td align="right">-0.1401523</td>
<td align="right">0.0079155</td>
</tr>
<tr class="odd">
<td align="left">Gene_88</td>
<td align="right">0.1234406</td>
<td align="right">0.0194714</td>
</tr>
<tr class="even">
<td align="left">Gene_174</td>
<td align="right">-0.1272991</td>
<td align="right">0.0159525</td>
</tr>
<tr class="odd">
<td align="left">Gene_189</td>
<td align="right">-0.1462271</td>
<td align="right">0.0055715</td>
</tr>
<tr class="even">
<td align="left">Gene_190</td>
<td align="right">-0.1443251</td>
<td align="right">0.0062277</td>
</tr>
<tr class="odd">
<td align="left">Gene_222</td>
<td align="right">0.1258673</td>
<td align="right">0.0171873</td>
</tr>
<tr class="even">
<td align="left">Gene_367</td>
<td align="right">0.1499230</td>
<td align="right">0.0044714</td>
</tr>
</tbody>
</table>
<div class="figure">
<img src="SmCCNet_Vignette_SingleOmics_files/figure-html/loadings1-1.png" alt="PC1 loading for each subnetwork feature." width="100%"><p class="caption">
PC1 loading for each subnetwork feature.
</p>
</div>
<div class="figure">
<img src="SmCCNet_Vignette_SingleOmics_files/figure-html/corheatmap-1.png" alt="Correlation heatmap for subnetwork features." width="100%"><p class="caption">
Correlation heatmap for subnetwork features.
</p>
</div>
<div class="figure">
<img src="SmCCNet_Vignette_SingleOmics_files/figure-html/adjheatmap-1.png" alt="Adjacency matrix heatmap for subnetwork features." width="100%"><p class="caption">
Adjacency matrix heatmap for subnetwork features.
</p>
</div>
<p>Figure <span class="math inline">\(\ref{fig:loadings1}\)</span> the
visualization of the PC loadings that represents the contribution of
each molecular features to the first NetSHy PC. In addition, there are
two network heatmaps based on (1) correlation matrix (Figure <span class="math inline">\(\ref{fig:corheatmap}\)</span>), and (2) adjacency
matrix (Figure <span class="math inline">\(\ref{fig:adjheatmap}\)</span>). Based on the
summarization table, genes 1,2,6, and 7 have relatively high correlation
with respect to phenotype. The PC loadings also confirm that genes
1,2,5,6,and 7 generally have PC contribution. From the correlation
heatmap, we do not observe associations between molecular features, but
for the adjacency matrix heatmap, we observe the higher connections
between genes 1,2,6, and 7.</p>
<div class="section level3">
<h3 id="step-vi-visualize-network-module">Step VI: Visualize network module<a class="anchor" aria-label="anchor" href="#step-vi-visualize-network-module"></a>
</h3>
<p>The initial approach to network visualization is facilitated through
our SmCCNet shinyApp, accessible at <a href="https://smccnet.shinyapps.io/smccnetnetwork/" class="external-link">https://smccnet.shinyapps.io/smccnetnetwork/</a>.
Upon obtaining a subnetwork file named â€˜size_a_net_b.Rdataâ€™, users can
upload it to the shinyApp. The platform provides various adjustable
visualization parameters, enabling users to tailor the network
visualization to their preferences.</p>
<p>An alternative way to visualize the final network module, we need to
download the Cytoscape software, and use the package RCy3 to visualize
the subnetwork generated from the network trimming step. In general,
since the network obtained through the PageRank trimming algorithm is
densely connected, and some of the edges may be false positive (meaning
that two nodes are not associated, but with higher edge values in the
adjacency matrix). Therefore, we use the correlation matrix to filter
out those weak network edges.</p>
<p>In the network visualization (Figure <span class="math inline">\(\ref{fig:netPlot}\)</span>), different colored
edges denote different directions of the association between two nodes,
where red or blue denotes a positive or negative association
respectively. The width of the edge represents the connection strength
between two nodes.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">RCy3</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r.igraph.org/" class="external-link">igraph</a></span><span class="op">)</span></span>
<span><span class="co"># load subnetwork data (example, user need to provide the directory)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="st">'ResultDirectory/size_a_net_b.Rdata'</span><span class="op">)</span></span>
<span><span class="va">M</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">M</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># correlation matrix for the subnetwork</span></span>
<span><span class="va">filter_index</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">correlation_sub</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="va">M_ind</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">correlation_sub</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="fl">1</span>, <span class="op">-</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">M_adj</span> <span class="op">&lt;-</span> <span class="va">M</span> <span class="op">*</span> <span class="va">M_ind</span></span>
<span><span class="va">M_adj</span><span class="op">[</span><span class="va">filter_index</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">M_adj</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span></span>
<span><span class="co"># network visualization through cytoscape</span></span>
<span><span class="va">graph</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/graph_from_adjacency_matrix.html" class="external-link">graph_from_adjacency_matrix</a></span><span class="op">(</span><span class="va">M_adj</span>, mode <span class="op">=</span> <span class="st">'undirected'</span>, </span>
<span>         weighted <span class="op">=</span> <span class="cn">TRUE</span>, diag <span class="op">=</span> <span class="cn">TRUE</span>, add.colnames <span class="op">=</span> <span class="cn">NULL</span>, add.rownames <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># define network node type and connectivity</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">type</span> <span class="op">&lt;-</span> <span class="va">sub_type</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">type</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">connectivity</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">M</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://r.igraph.org/reference/V.html" class="external-link">V</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span><span class="op">$</span><span class="va">connectivity</span></span>
<span></span>
<span><span class="fu">createNetworkFromIgraph</span><span class="op">(</span><span class="va">graph</span>,<span class="st">"single_omics_network"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="../articles/example_network_cont.png" alt="Trimmed module 1. The strength of the node connections is indicated by the thickness of edges. Red edges and blue edges are for negative and positive connections respectively." width="100%"><p class="caption">
Trimmed module 1. The strength of the node connections is indicated by
the thickness of edges. Red edges and blue edges are for negative and
positive connections respectively.
</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="single-omics-smccnet-for-binary-phenotype">Single-omics SmCCNet for Binary Phenotype<a class="anchor" aria-label="anchor" href="#single-omics-smccnet-for-binary-phenotype"></a>
</h2>
<p>This section provides a tutorial on running the SmCCNet algorithm for
a binary phenotype using single-omics data. Given the binary nature of
the phenotype, canonical correlation analysis can be applied, but it may
pose challenges due to the correlation issues with binary random
variables. To address this, we employ Sparse Partial Least Squares
Discriminant Analysis (SPLSDA) to modify the SmCCNet algorithm for
binary phenotype.</p>
<p>SPLSDA operates with a dual-stage approach aimed at predicting the
phenotype. Initially, it treats the binary phenotype as a quantitative
variable and applies partial least squares to derive multiple latent
factors. Subsequently, these latent factors are used to predict the
binary phenotype using logistic regression. The latent factors are then
consolidated into a single factor based on the logistic regression
coefficient.</p>
<p>Similar to sparse multiple canonical correlation analysis (SmCCA)
used in the quantitative phenotype analysis for SmCCNet, a sparsity
parameter needs to be tuned. However, for SPLSDA, an increase in the
penalty value results in a higher level of sparsity, which is contrary
to the CCA-based method implemented in this package. Itâ€™s also important
to note that this algorithm is designed specifically for binary
phenotypes. For users dealing with multi-class phenotypes, a
one-versus-all approach can be used to generate networks separately if
the multi-class phenotype is not ordinal. Alternatively, if the
multi-class phenotype is ordinal, a CCA-based approach can be
applied.</p>
<div class="section level3">
<h3 id="import-synthetic-dataset-with-binary-phenotype">Import synthetic dataset with binary phenotype<a class="anchor" aria-label="anchor" href="#import-synthetic-dataset-with-binary-phenotype"></a>
</h3>
<p>As an example, we consider a synthetic data set with 500 genes <span class="math inline">\((X_1)\)</span> expression levels measured for 358
subjects, along with a binarized phenotype <span class="math inline">\((Y)\)</span> by taking the median of original
phenotype as the cutoff, and transforming it into binary variable.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">ExampleData</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">X1</span><span class="op">[</span> , <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          Gene_1   Gene_2   Gene_3   Gene_4   Gene_5   Gene_6</span></span>
<span><span class="co">## Samp_1 22.48570 40.35372 31.02575 20.84721 26.69729 30.20545</span></span>
<span><span class="co">## Samp_2 37.05885 34.05223 33.48702 23.53146 26.75463 31.73594</span></span>
<span><span class="co">## Samp_3 20.53077 31.66962 35.18957 20.95254 25.01883 32.15723</span></span>
<span><span class="co">## Samp_4 33.18689 38.48088 18.89710 31.82330 34.04938 38.79989</span></span>
<span><span class="co">## Samp_5 28.96198 41.06049 28.49496 18.37449 30.81524 24.00454</span></span>
<span><span class="co">## Samp_6 18.05983 29.55471 32.54002 29.68452 26.19996 26.76684</span></span></code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##           Pheno</span></span>
<span><span class="co">## Samp_1 235.0674</span></span>
<span><span class="co">## Samp_2 253.5450</span></span>
<span><span class="co">## Samp_3 234.2050</span></span>
<span><span class="co">## Samp_4 281.0354</span></span>
<span><span class="co">## Samp_5 245.4478</span></span>
<span><span class="co">## Samp_6 189.6231</span></span></code></pre>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># binarize quantitative outcome </span></span>
<span><span class="va">Y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">Y</span> <span class="op">&gt;</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span>, <span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p>Denote the number of features in <span class="math inline">\(X_1\)</span> as <span class="math inline">\(p_1\)</span>, and the number of subjects as <span class="math inline">\(n\)</span>.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">X1</span><span class="op">)</span></span>
<span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">X1</span><span class="op">)</span></span>
<span><span class="va">AbarLabel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">X1</span><span class="op">)</span></span></code></pre></div>
<p>The parameter setup for this approach mirrors that of SmCCA, and for
more specific information, please refer back to the single-omics
quantitative phenotype in section 1.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">K</span> <span class="op">&lt;-</span> <span class="fl">3</span> <span class="co"># number of folds in K-fold CV.</span></span>
<span><span class="va">s1</span> <span class="op">&lt;-</span> <span class="fl">0.7</span> <span class="co"># feature subsampling percentage.</span></span>
<span><span class="va">subSamp</span> <span class="op">&lt;-</span> <span class="fl">50</span> <span class="co"># number of subsamples.</span></span>
<span><span class="va">CCcoef</span> <span class="op">&lt;-</span> <span class="cn">NULL</span> <span class="co"># unweighted version of SmCCNet.</span></span>
<span><span class="va">metric</span> <span class="op">&lt;-</span> <span class="st">'auc'</span> <span class="co"># evaluation metric to be used.</span></span>
<span></span>
<span><span class="co"># create sparsity penalty options.</span></span>
<span><span class="va">pen1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">.1</span>, <span class="fl">.9</span>, by <span class="op">=</span> <span class="fl">.1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># set a CV directory.</span></span>
<span><span class="va">CVDir</span> <span class="op">&lt;-</span> <span class="st">"Example3foldCV_Binary/"</span> </span>
<span><span class="va">pheno</span> <span class="op">&lt;-</span> <span class="st">"Example3foldCV_Binary"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">CVDir</span><span class="op">)</span></span>
<span></span>
<span><span class="va">Y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">Y</span> <span class="op">&gt;</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span>, <span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="step-ii-determine-optimal-sparsity-penalty-through-cross-validation-optional">Step II: Determine optimal sparsity penalty through cross-validation
(optional)<a class="anchor" aria-label="anchor" href="#step-ii-determine-optimal-sparsity-penalty-through-cross-validation-optional"></a>
</h3>
<p>The <strong>SPLS</strong> package does not include a method for
assessing prediction accuracy. Therefore, we utilize the glm() function
from the <strong>stats</strong> package to evaluate the performance of
our model. Particularly for SPLSDA, the number of components is an
important factor to consider as the goal is to make predictions. In the
following example, we select three as the number of components. All
other procedures remain identical to those in SmCCA (see single-omics
quantitative phenotype vignette).</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">12345</span><span class="op">)</span> <span class="co"># set random seed.</span></span>
<span><span class="co"># save original data into local directory</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span><span class="va">X1</span>, <span class="va">Y</span>, <span class="va">s1</span>, <span class="va">subSamp</span>, <span class="va">pen1</span>, </span>
<span>     file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">CVDir</span>, <span class="st">"Data.Rdata"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># define the number of components to be extracted</span></span>
<span><span class="va">ncomp</span> <span class="op">&lt;-</span> <span class="fl">3</span></span>
<span><span class="co"># split data into k folds</span></span>
<span><span class="va">foldIdx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">N</span>, <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">N</span>, <span class="va">K</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">K</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">iIdx</span> <span class="op">&lt;-</span> <span class="va">foldIdx</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="va">x1.train</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/scale.html" class="external-link">scale</a></span><span class="op">(</span><span class="va">X1</span><span class="op">[</span><span class="op">-</span><span class="va">iIdx</span>, <span class="op">]</span><span class="op">)</span></span>
<span>  <span class="va">yy.train</span> <span class="op">&lt;-</span> <span class="va">Y</span><span class="op">[</span><span class="op">-</span><span class="va">iIdx</span>, <span class="op">]</span></span>
<span>  <span class="va">x1.test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/scale.html" class="external-link">scale</a></span><span class="op">(</span><span class="va">X1</span><span class="op">[</span><span class="va">iIdx</span>, <span class="op">]</span><span class="op">)</span></span>
<span>  <span class="va">yy.test</span> <span class="op">&lt;-</span> <span class="va">Y</span><span class="op">[</span><span class="va">iIdx</span>, <span class="op">]</span></span>
<span>  </span>
<span>  <span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">x1.train</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">yy.train</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">x1.test</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">yy.test</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/stop.html" class="external-link">stop</a></span><span class="op">(</span><span class="st">"Invalid scaled data."</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>  <span class="va">subD</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">CVDir</span>, <span class="st">"CV_"</span>, <span class="va">i</span>, <span class="st">"/"</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">subD</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span><span class="va">x1.train</span>, <span class="va">yy.train</span>, <span class="va">x1.test</span>, <span class="va">yy.test</span>,</span>
<span>      <span class="va">pen1</span>, <span class="va">p1</span>, </span>
<span>       file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">subD</span>, <span class="st">"Data.Rdata"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co">############################################## Running Cross-Validation</span></span>
<span></span>
<span><span class="co"># run SmCCA with K-fold cross-validation</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="kw">for</span> <span class="op">(</span><span class="va">CVidx</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">K</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="co"># define evaluation method</span></span>
<span>  <span class="va">EvalMethod</span> <span class="op">&lt;-</span> <span class="st">'precision'</span></span>
<span>  <span class="co"># define and create saving directory</span></span>
<span>  <span class="va">subD</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">CVDir</span>, <span class="st">"CV_"</span>, <span class="va">CVidx</span>, <span class="st">"/"</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">subD</span>, <span class="st">"Data.Rdata"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">subD</span>, <span class="st">"SmCCA/"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># create empty vector to store cross-validation accuracy result</span></span>
<span>  <span class="va">TrainScore</span> <span class="op">&lt;-</span> <span class="va">TestScore</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">pen1</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="kw">for</span><span class="op">(</span><span class="va">idx</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">pen1</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="co"># define value of penalty</span></span>
<span>    <span class="va">l1</span> <span class="op">&lt;-</span> <span class="va">pen1</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span></span>
<span></span>
<span>    <span class="co"># run SPLS-DA to extract canonical weight</span></span>
<span>    <span class="va">Ws</span> <span class="op">&lt;-</span> <span class="fu">spls</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/spls/man/splsda.html" class="external-link">splsda</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x1.train</span>, y <span class="op">=</span> <span class="va">yy.train</span>, K <span class="op">=</span> <span class="va">ncomp</span>, eta <span class="op">=</span> <span class="va">l1</span>, </span>
<span>                       kappa<span class="op">=</span><span class="fl">0.5</span>,</span>
<span>                       classifier<span class="op">=</span> <span class="st">'logistic'</span>, scale.x<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co"># create emtpy matrix to save canonical weights for each subsampling</span></span>
<span>    <span class="va">weight</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">0</span>,nrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">x1.train</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="va">ncomp</span><span class="op">)</span></span>
<span>    <span class="va">weight</span><span class="op">[</span><span class="va">Ws</span><span class="op">[[</span><span class="st">"A"</span><span class="op">]</span><span class="op">]</span>,<span class="op">]</span> <span class="op">&lt;-</span> <span class="va">Ws</span><span class="op">[[</span><span class="st">"W"</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span>    <span class="co"># train the latent factor model with logistic regression</span></span>
<span>    <span class="va">train_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>x <span class="op">=</span> <span class="op">(</span><span class="va">x1.train</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span> <span class="va">weight</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="va">ncomp</span><span class="op">]</span>, </span>
<span>                             y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">yy.train</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">test_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>x <span class="op">=</span> <span class="op">(</span><span class="va">x1.test</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span> <span class="va">weight</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="va">ncomp</span><span class="op">]</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="va">logisticFit</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/glm.html" class="external-link">glm</a></span><span class="op">(</span><span class="va">y</span><span class="op">~</span><span class="va">.</span>, family <span class="op">=</span> <span class="st">'binomial'</span>,data <span class="op">=</span> <span class="va">train_data</span><span class="op">)</span></span>
<span>    <span class="co"># make prediction for train/test set</span></span>
<span>    <span class="va">train_pred</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">logisticFit</span>, <span class="va">train_data</span>, type <span class="op">=</span> <span class="st">'response'</span><span class="op">)</span></span>
<span>    <span class="va">test_pred</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">logisticFit</span>, <span class="va">test_data</span>, type <span class="op">=</span> <span class="st">'response'</span><span class="op">)</span></span>
<span>    <span class="co"># specifying which method to use as evaluation metric</span></span>
<span>    <span class="va">TrainScore</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/classifierEval.html">classifierEval</a></span><span class="op">(</span>obs <span class="op">=</span> <span class="va">yy.train</span>, </span>
<span>                                      pred <span class="op">=</span> <span class="va">train_pred</span>, </span>
<span>                                      EvalMethod <span class="op">=</span> <span class="va">metric</span>, print_score <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span>    <span class="va">TestScore</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/classifierEval.html">classifierEval</a></span><span class="op">(</span>obs <span class="op">=</span> <span class="va">yy.test</span>, </span>
<span>                                      pred <span class="op">=</span> <span class="va">test_pred</span>, </span>
<span>                                      EvalMethod <span class="op">=</span> <span class="va">metric</span>, print_score <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span>    </span>
<span>  <span class="op">}</span></span>
<span></span>
<span>  <span class="co"># combine cross-validation results</span></span>
<span>  <span class="va">DeltaAcc.all</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">pen1</span>, <span class="va">TrainScore</span>, <span class="va">TestScore</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">DeltaAcc.all</span><span class="op">$</span><span class="va">Delta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">DeltaAcc.all</span><span class="op">$</span><span class="va">TrainScore</span> <span class="op">-</span> <span class="va">DeltaAcc.all</span><span class="op">$</span><span class="va">TestScore</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">DeltaAcc.all</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"l1"</span>, <span class="st">"Training Score"</span>, <span class="st">"Testing Score"</span>, <span class="st">"Delta"</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># save cross-validation results to local directory</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.csv</a></span><span class="op">(</span><span class="va">DeltaAcc.all</span>,</span>
<span>            file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">subD</span>, <span class="st">"SmCCA/SCCA_"</span>, <span class="va">subSamp</span>,<span class="st">"_allDeltaCor.csv"</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span><span class="op">}</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="step-iii-run-splsda-with-pre-selected-penalty-term">Step III: Run SPLSDA with pre-selected penalty term<a class="anchor" aria-label="anchor" href="#step-iii-run-splsda-with-pre-selected-penalty-term"></a>
</h3>
<p>The evaluation metric is maximizing the classification evaluation
metric such as testing prediction accuracy or testing AUC score.
Currently there are 5 different evaluation metrics to choose from:
accuracy, AUC score, precision, recall, and F1 score, with accuracy as
the default selection (see multi-omics vignette section 5 for details).
The following example employs the penalty term with the greatest testing
prediction accuracy to execute the SPLSDA algorithm on the entire
dataset.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># save cross-validation result</span></span>
<span><span class="va">cv_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggregateCVSingle.html">aggregateCVSingle</a></span><span class="op">(</span><span class="va">CVDir</span>, <span class="st">"SmCCA"</span>, NumSubsamp <span class="op">=</span> <span class="va">subSamp</span>, K <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># create directory to save overall result with the best penalty term</span></span>
<span><span class="va">plotD</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">CVDir</span>, <span class="st">"Figures/"</span><span class="op">)</span></span>
<span><span class="va">saveD</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">CVDir</span>, <span class="st">"Results/"</span><span class="op">)</span></span>
<span><span class="va">dataF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">CVDir</span>, <span class="st">"Data.Rdata"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">plotD</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">saveD</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">dataF</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># specify method (only SmCCA works)</span></span>
<span><span class="va">Method</span> <span class="op">=</span> <span class="st">"SmCCA"</span></span>
<span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">Method</span> <span class="kw">in</span> <span class="st">"SmCCA"</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="co"># read cross validation result in</span></span>
<span>  <span class="va">T12</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">CVDir</span>, <span class="st">"Results/"</span>, <span class="va">Method</span>, <span class="st">"CVmeanDeltaCors.csv"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="co"># determine the optimal penalty term</span></span>
<span>  <span class="va">pen</span> <span class="op">&lt;-</span> <span class="va">T12</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.min.html" class="external-link">which.max</a></span><span class="op">(</span><span class="va">T12</span><span class="op">[</span> ,<span class="fl">3</span><span class="op">]</span><span class="op">)</span> ,<span class="fl">2</span><span class="op">]</span></span>
<span>  <span class="va">l1</span> <span class="op">&lt;-</span> <span class="va">pen</span>;</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="co"># run SPLSDA with optimal penalty term</span></span>
<span>    <span class="va">Ws</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getRobustWeightsSingleBinary.html">getRobustWeightsSingleBinary</a></span><span class="op">(</span>X1 <span class="op">=</span> <span class="va">X1</span>, Trait <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span>, </span>
<span>                                       Lambda1 <span class="op">=</span> <span class="va">l1</span>, </span>
<span>                                       <span class="va">s1</span>, SubsamplingNum <span class="op">=</span> <span class="va">subSamp</span><span class="op">)</span></span>
<span>   </span>
<span>    <span class="co"># get adjacency matrix</span></span>
<span>    <span class="va">Abar</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getAbar.html">getAbar</a></span><span class="op">(</span><span class="va">Ws</span>, FeatureLabel <span class="op">=</span> <span class="va">AbarLabel</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">p1</span><span class="op">]</span><span class="op">)</span></span>
<span>    <span class="co"># save result into local directory</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span><span class="va">l1</span>, <span class="va">X1</span>, <span class="va">Y</span>, <span class="va">s1</span>, <span class="va">Ws</span>, <span class="va">Abar</span>,</span>
<span>         file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">saveD</span>, <span class="va">Method</span>, <span class="va">K</span>, <span class="st">"foldSamp"</span>, <span class="va">subSamp</span>, <span class="st">"_"</span>, <span class="va">pen</span>, </span>
<span>                       <span class="st">".Rdata"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span><span class="op">)</span></span>
<span>  </span>
<span>  </span>
<span><span class="op">}</span></span></code></pre></div>
<p>After this step, all the other steps are a consistent with
single-omics quantitative phenotype example, refer to section 1.4 and
1.5 for more information.</p>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.2 (2023-10-31)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] grid      parallel  stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">## [8] methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] reshape2_1.4.4   shadowtext_0.1.2 lubridate_1.9.3  forcats_1.0.0   </span></span>
<span><span class="co">##  [5] stringr_1.5.1    dplyr_1.1.4      purrr_1.0.2      readr_2.1.5     </span></span>
<span><span class="co">##  [9] tidyr_1.3.0      tibble_3.2.1     ggplot2_3.4.4    tidyverse_2.0.0 </span></span>
<span><span class="co">## [13] SmCCNet_2.0.2    igraph_1.6.0     Matrix_1.6-1.1   pbapply_1.7-2   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] sass_0.4.8        utf8_1.2.4        generics_0.1.3    stringi_1.8.3    </span></span>
<span><span class="co">##  [5] lattice_0.21-9    hms_1.1.3         digest_0.6.34     magrittr_2.0.3   </span></span>
<span><span class="co">##  [9] timechange_0.2.0  evaluate_0.23     fastmap_1.1.1     plyr_1.8.9       </span></span>
<span><span class="co">## [13] jsonlite_1.8.8    fansi_1.0.6       scales_1.3.0      textshaping_0.3.7</span></span>
<span><span class="co">## [17] jquerylib_0.1.4   cli_3.6.2         rlang_1.1.3       munsell_0.5.0    </span></span>
<span><span class="co">## [21] withr_2.5.2       cachem_1.0.8      yaml_2.3.8        tools_4.3.2      </span></span>
<span><span class="co">## [25] tzdb_0.4.0        memoise_2.0.1     colorspace_2.1-0  vctrs_0.6.5      </span></span>
<span><span class="co">## [29] R6_2.5.1          lifecycle_1.0.4   fs_1.6.3          ragg_1.2.7       </span></span>
<span><span class="co">## [33] pkgconfig_2.0.3   desc_1.4.3        pkgdown_2.0.7     bslib_0.6.1      </span></span>
<span><span class="co">## [37] pillar_1.9.0      gtable_0.3.4      Rcpp_1.0.12       glue_1.7.0       </span></span>
<span><span class="co">## [41] systemfonts_1.0.5 xfun_0.41         tidyselect_1.2.0  highr_0.10       </span></span>
<span><span class="co">## [45] knitr_1.45        farver_2.1.1      htmltools_0.5.7   labeling_0.4.3   </span></span>
<span><span class="co">## [49] rmarkdown_2.25    compiler_4.3.2</span></span></code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/warnings.html" class="external-link">warnings</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<blockquote>
<p>Shi, W. J., Zhuang, Y., Russell, P. H., Hobbs, B. D., Parker, M. M.,
Castaldi, P. J., â€¦ &amp; Kechris, K. (2019). Unsupervised discovery of
phenotype-specific multi-omics networks. Bioinformatics, 35(21),
4336-4343.</p>
</blockquote>
<blockquote>
<p>Vu, T., Litkowski, E. M., Liu, W., Pratte, K. A., Lange, L., Bowler,
R. P., â€¦ &amp; Kechris, K. J. (2023). NetSHy: network summarization via
a hybrid approach leveraging topological properties. Bioinformatics,
39(1), btac818.</p>
</blockquote>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Weixuan Liu.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
